{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FlashNet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use Tencent I/O traces that were replayed on our NVMe device. This data is ready to be used for training after applying some preprocessing. As for the raw version of Tencent trace, it can be downloaded here: https://www.googleapis.com/drive/v3/files/1Xj6rFBOsY9Wt_XlCiAiZn7LfSCAYy1AB?alt=media&key=AIzaSyBbUdTut1W8uzPO1nzCmcHFIw0KsuO3Dfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Goal\n",
    "This notebook will guide you through the process of preparing a trace profile to train a Neural Network (NN) model for performing per-IO admission control."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup Connection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to Pre-made Lease"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**How to create Lease at Chameleon Cloud?**\n",
    "\n",
    "* Create Reservation <br>\n",
    "    Go to https://chi.tacc.chameleoncloud.org/project/ or https://chi.uc.chameleoncloud.org/project/\n",
    "    \n",
    "    Reserve Physical Host\n",
    "    - click \"Leases\" => \"+ Create Lease\"\n",
    "    - lease name = \"trovi-ceph\"\n",
    "    - \"$node_type\", \"storage_nvme\"\n",
    "    - Max lease: 7 days\n",
    "\n",
    "* Launching an Instance\n",
    "    - In the sidebar, click Compute, then click Instances\n",
    "    - Click on the Launch Instance\n",
    "        + pick the correct reservation \n",
    "        + count = 1\n",
    "        + Image: CC-Ubuntu20.04\n",
    "    - Choose the ssh key\n",
    "\n",
    "* Allocate floating IPs\n",
    "    - Book the IP interface\n",
    "        + Click \"Network -> Floating IPs -> Allocate IP To Project\"\n",
    "        + Write description (optional)\n",
    "        + Click \"Allocate IP\"\n",
    "    - Click \"Associate\" OR click \"attach interface\"\n",
    "        + Click \"Network -> Floating IPs\"\n",
    "    - Wait a few minutes until node spawned\n",
    "\n",
    "* If you can't access the cc user on SSH\n",
    "    - open the console terminal via the website\n",
    "    - edit the .ssh/authorized_keys and add your pub_key manually!!!\n",
    "    - now, the hostname will be different \n",
    "    - run \"sudo dhclient\" from the web-based ssh console"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Declare basic variables for ssh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "export ip_addr=\"192.5.86.149\"\n",
    "export SSHKEY_NAME=\"$USER-flashnet\"\n",
    "export SSHKEY_FILE=\"$HOME/.ssh/$SSHKEY_NAME\"\n",
    "export USER=\"$USER\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create ssh folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ~/.ssh\n",
    "echo \"Host *\" > ~/.ssh/config\n",
    "echo \" StrictHostKeyChecking no\" >> ~/.ssh/config\n",
    "sudo rm -rf output/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create ssh key pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating public/private rsa key pair.\n",
      "Your identification has been saved in /home/rani_api3939_gmail_com/.ssh/rani_api3939_gmail_com-flashnet\n",
      "Your public key has been saved in /home/rani_api3939_gmail_com/.ssh/rani_api3939_gmail_com-flashnet.pub\n",
      "The key fingerprint is:\n",
      "SHA256:CGqbJS2f1xOtaLNCkLAiB3pjit8KQJUqudvgaqUZvMc rani_api3939_gmail_com-flashnet\n",
      "The key's randomart image is:\n",
      "+---[RSA 4096]----+\n",
      "|   ..            |\n",
      "|o ..             |\n",
      "|.*...            |\n",
      "|O.Bo . . .       |\n",
      "|B*=oo . S .      |\n",
      "|*+ O.. o o       |\n",
      "|o=Xoo = +        |\n",
      "|.BoEoo o .       |\n",
      "|+ o. ..          |\n",
      "+----[SHA256]-----+\n",
      "SSH key with name: rani_api3939_gmail_com-flashnet created\n"
     ]
    }
   ],
   "source": [
    "rm -rf $HOME/.ssh/$USER-*\n",
    "ssh-keygen -t rsa -b 4096  -P '' -C $SSHKEY_NAME -f $SSHKEY_FILE <<< y\n",
    "echo \"SSH key with name: $SSHKEY_NAME created\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add this Trovi session machine's public key to reserved node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "echo \"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQC6yeckn+S3SJsTgdSL48tQkp/QkASNoWNx+U7zdHfatsX1x5FGjWtx4x+yIq4rEBXLr703WiV4zGzZSiji6ywTlDBxIyJqpQvlDpflR7xRBBKPwyfu/TDiMXpPRORfoXtLdp9P/DfY64h5rs/6G1MZd+shH6QXJtSrxhTzc3vcGSK1xBvGHIaDVWZ01Rna3QbpzB3H7kurdu0Ku4F2cShVybRFfGgg+QCcuQLb6vtFk+2uPJslj7tY0bZiLA0WJBq6Zw998YXfaLkH3SP3+nE9Ryy62/xGpM4Itzn6nC0ao5pK7nSfPAtzzpOxfbM4XkAK6Pa6IM+qT0ZwO8rlkaOcVc2ALmPvWTdw8fBvRkL/+AUAPKrgZ5yAUCSUBII5CXZoNTnYaUr1ddkVmtTyXx9+gRZY5oci1jODAPBpTV+zvyX6zVMqMUQ7JuKkM1IpQY/bcCgL6163L5HRKabdjli1s8EgWrLfqBvjQ2dsYA898KyK9KSFUpxcWm/s9KVb8ZA2JKQs8CeLO2DBS9rs3IDmwMe1wX3NVO7yMWgKTzm/PJVMzeDAwoWjIdknW7UF267ERttuVRApqVzUHQ/02QpI3Q+mLxneS4pZh2OKMrnx4SDY+UfNzxR5ZIMh+DIQ/+AtCMPFl+oi++G6aIOvJkXpMQK28o03w9kf5oDUVhuw1w== rani_api3939_gmail_com-flashnet\" >> ~/.ssh/authorized_keys\n",
      "chmod 644 ~/.ssh/authorized_keys\n",
      "echo 'StrictHostKeyChecking no' > ~/.ssh/config; chmod go-rw ~/.ssh/config\n"
     ]
    }
   ],
   "source": [
    "pubkey=`cat $SSHKEY_FILE.pub`\n",
    "echo \"echo \\\"$pubkey\\\" >> ~/.ssh/authorized_keys\"\n",
    "echo \"chmod 644 ~/.ssh/authorized_keys\"\n",
    "echo \"echo 'StrictHostKeyChecking no' > ~/.ssh/config; chmod go-rw ~/.ssh/config\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the output of above cell to the reserved node.\n",
    "\n",
    "Steps:\n",
    "1. Open your local terminal\n",
    "2. Connect to the node (ssh cc@xxx.xx.xx)\n",
    "3. Copy the output of cell above and paste to terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check Connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cc\n",
      "Waiting up to 300 seconds for SSH on 192.5.86.149...\n",
      "SSH is running!\n"
     ]
    }
   ],
   "source": [
    "ssh_command=\"ssh -o \\\"StrictHostKeyChecking no\\\" -i $SSHKEY_FILE cc@$ip_addr\"\n",
    "eval \"$ssh_command\" pwd \n",
    "wait_ssh \"$ip_addr\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If a specific node is used, then for some reasons rebuilding the node is needed, **run this cell after the node has been rebuilt**. Otherwise, ignore."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    ssh-keygen -f \"/home/$USER/.ssh/known_hosts\" -R $ip_addr\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step-by-step Guideline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'flashnet-trovi'...\n",
      "/home/cc/flashnet-trovi\n"
     ]
    }
   ],
   "source": [
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    git clone https://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@github.com/daniarherikurniawan/flashnet-trovi.git\n",
    "    cd flashnet-trovi\n",
    "    find . -type f -iname \"*.sh\" -exec chmod +x {} \\;\n",
    "    find . -type f -iname \"*.py\" -exec chmod +x {} \\;\n",
    "    pwd\n",
    "    sudo chown cc -R .\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Install Conda Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anaconda has already installed.\n",
      "no change     /home/cc/anaconda3/condabin/conda\n",
      "no change     /home/cc/anaconda3/bin/conda\n",
      "no change     /home/cc/anaconda3/bin/conda-env\n",
      "no change     /home/cc/anaconda3/bin/activate\n",
      "no change     /home/cc/anaconda3/bin/deactivate\n",
      "no change     /home/cc/anaconda3/etc/profile.d/conda.sh\n",
      "no change     /home/cc/anaconda3/etc/fish/conf.d/conda.fish\n",
      "no change     /home/cc/anaconda3/shell/condabin/Conda.psm1\n",
      "no change     /home/cc/anaconda3/shell/condabin/conda-hook.ps1\n",
      "no change     /home/cc/anaconda3/lib/python3.9/site-packages/xontrib/conda.xsh\n",
      "no change     /home/cc/anaconda3/etc/profile.d/conda.csh\n",
      "no change     /home/cc/.zshrc\n",
      "No action taken.\n",
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.9.0\n",
      "  latest version: 23.1.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/cc/anaconda3/envs/flashnet-trovi-env\n",
      "\n",
      "  added / updated specs:\n",
      "    - python=3.8\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main None\n",
      "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-5.1-1_gnu None\n",
      "  ca-certificates    pkgs/main/linux-64::ca-certificates-2023.01.10-h06a4308_0 None\n",
      "  certifi            pkgs/main/linux-64::certifi-2022.12.7-py38h06a4308_0 None\n",
      "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.38-h1181459_1 None\n",
      "  libffi             pkgs/main/linux-64::libffi-3.4.2-h6a678d5_6 None\n",
      "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-11.2.0-h1234567_1 None\n",
      "  libgomp            pkgs/main/linux-64::libgomp-11.2.0-h1234567_1 None\n",
      "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-11.2.0-h1234567_1 None\n",
      "  ncurses            pkgs/main/linux-64::ncurses-6.4-h6a678d5_0 None\n",
      "  openssl            pkgs/main/linux-64::openssl-1.1.1s-h7f8727e_0 None\n",
      "  pip                pkgs/main/linux-64::pip-22.3.1-py38h06a4308_0 None\n",
      "  python             pkgs/main/linux-64::python-3.8.16-h7a1cb2a_2 None\n",
      "  readline           pkgs/main/linux-64::readline-8.2-h5eee18b_0 None\n",
      "  setuptools         pkgs/main/linux-64::setuptools-65.6.3-py38h06a4308_0 None\n",
      "  sqlite             pkgs/main/linux-64::sqlite-3.40.1-h5082296_0 None\n",
      "  tk                 pkgs/main/linux-64::tk-8.6.12-h1ccaba5_0 None\n",
      "  wheel              pkgs/main/noarch::wheel-0.37.1-pyhd3eb1b0_0 None\n",
      "  xz                 pkgs/main/linux-64::xz-5.2.10-h5eee18b_1 None\n",
      "  zlib               pkgs/main/linux-64::zlib-1.2.13-h5eee18b_0 None\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate flashnet-trovi-env\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "Retrieving notices: ...working... done\n",
      "/home/cc/anaconda3/envs/flashnet-trovi-env/bin/pip\n",
      "pip 22.3.1 from /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages/pip (python 3.8)\n",
      "Collecting tensorflow==2.7.0\n",
      "  Using cached tensorflow-2.7.0-cp38-cp38-manylinux2010_x86_64.whl (489.6 MB)\n",
      "Collecting tensorflow-estimator<2.8,~=2.7.0rc0\n",
      "  Using cached tensorflow_estimator-2.7.0-py2.py3-none-any.whl (463 kB)\n",
      "Collecting protobuf>=3.9.2\n",
      "  Using cached protobuf-4.21.12-cp37-abi3-manylinux2014_x86_64.whl (409 kB)\n",
      "Collecting h5py>=2.9.0\n",
      "  Using cached h5py-3.8.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.7 MB)\n",
      "Collecting six>=1.12.0\n",
      "  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)\n",
      "Collecting grpcio<2.0,>=1.24.3\n",
      "  Using cached grpcio-1.51.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.8 MB)\n",
      "Collecting astunparse>=1.6.0\n",
      "  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)\n",
      "Collecting keras<2.8,>=2.7.0rc0\n",
      "  Using cached keras-2.7.0-py2.py3-none-any.whl (1.3 MB)\n",
      "Collecting absl-py>=0.4.0\n",
      "  Using cached absl_py-1.4.0-py3-none-any.whl (126 kB)\n",
      "Collecting libclang>=9.0.1\n",
      "  Using cached libclang-15.0.6.1-py2.py3-none-manylinux2010_x86_64.whl (21.5 MB)\n",
      "Collecting typing-extensions>=3.6.6\n",
      "  Using cached typing_extensions-4.4.0-py3-none-any.whl (26 kB)\n",
      "Collecting google-pasta>=0.1.1\n",
      "  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)\n",
      "Collecting flatbuffers<3.0,>=1.12\n",
      "  Using cached flatbuffers-2.0.7-py2.py3-none-any.whl (26 kB)\n",
      "Collecting gast<0.5.0,>=0.2.1\n",
      "  Using cached gast-0.4.0-py3-none-any.whl (9.8 kB)\n",
      "Collecting numpy>=1.14.5\n",
      "  Using cached numpy-1.24.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)\n",
      "Requirement already satisfied: wheel<1.0,>=0.32.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorflow==2.7.0) (0.37.1)\n",
      "Collecting opt-einsum>=2.3.2\n",
      "  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)\n",
      "Collecting wrapt>=1.11.0\n",
      "  Using cached wrapt-1.14.1-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (81 kB)\n",
      "Collecting termcolor>=1.1.0\n",
      "  Using cached termcolor-2.2.0-py3-none-any.whl (6.6 kB)\n",
      "Collecting keras-preprocessing>=1.1.1\n",
      "  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)\n",
      "Collecting tensorboard~=2.6\n",
      "  Using cached tensorboard-2.11.2-py3-none-any.whl (6.0 MB)\n",
      "Collecting tensorflow-io-gcs-filesystem>=0.21.0\n",
      "  Using cached tensorflow_io_gcs_filesystem-0.30.0-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (2.4 MB)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard~=2.6->tensorflow==2.7.0) (65.6.3)\n",
      "Collecting requests<3,>=2.21.0\n",
      "  Using cached requests-2.28.2-py3-none-any.whl (62 kB)\n",
      "Collecting markdown>=2.6.8\n",
      "  Using cached Markdown-3.4.1-py3-none-any.whl (93 kB)\n",
      "Collecting tensorboard-data-server<0.7.0,>=0.6.0\n",
      "  Using cached tensorboard_data_server-0.6.1-py3-none-manylinux2010_x86_64.whl (4.9 MB)\n",
      "Collecting google-auth-oauthlib<0.5,>=0.4.1\n",
      "  Using cached google_auth_oauthlib-0.4.6-py2.py3-none-any.whl (18 kB)\n",
      "Collecting google-auth<3,>=1.6.3\n",
      "  Using cached google_auth-2.16.0-py2.py3-none-any.whl (177 kB)\n",
      "Collecting tensorboard-plugin-wit>=1.6.0\n",
      "  Using cached tensorboard_plugin_wit-1.8.1-py3-none-any.whl (781 kB)\n",
      "Collecting werkzeug>=1.0.1\n",
      "  Using cached Werkzeug-2.2.2-py3-none-any.whl (232 kB)\n",
      "Collecting protobuf>=3.9.2\n",
      "  Using cached protobuf-3.20.3-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.0 MB)\n",
      "Collecting pyasn1-modules>=0.2.1\n",
      "  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)\n",
      "Collecting cachetools<6.0,>=2.0.0\n",
      "  Using cached cachetools-5.3.0-py3-none-any.whl (9.3 kB)\n",
      "Collecting rsa<5,>=3.1.4\n",
      "  Using cached rsa-4.9-py3-none-any.whl (34 kB)\n",
      "Collecting requests-oauthlib>=0.7.0\n",
      "  Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)\n",
      "Collecting importlib-metadata>=4.4\n",
      "  Using cached importlib_metadata-6.0.0-py3-none-any.whl (21 kB)\n",
      "Collecting urllib3<1.27,>=1.21.1\n",
      "  Using cached urllib3-1.26.14-py2.py3-none-any.whl (140 kB)\n",
      "Collecting charset-normalizer<4,>=2\n",
      "  Using cached charset_normalizer-3.0.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (195 kB)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow==2.7.0) (2022.12.7)\n",
      "Collecting idna<4,>=2.5\n",
      "  Using cached idna-3.4-py3-none-any.whl (61 kB)\n",
      "Collecting MarkupSafe>=2.1.1\n",
      "  Using cached MarkupSafe-2.1.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)\n",
      "Collecting zipp>=0.5\n",
      "  Using cached zipp-3.12.0-py3-none-any.whl (6.6 kB)\n",
      "Collecting pyasn1<0.5.0,>=0.4.6\n",
      "  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)\n",
      "Collecting oauthlib>=3.0.0\n",
      "  Using cached oauthlib-3.2.2-py3-none-any.whl (151 kB)\n",
      "Installing collected packages: tensorflow-estimator, tensorboard-plugin-wit, pyasn1, libclang, keras, flatbuffers, charset-normalizer, zipp, wrapt, urllib3, typing-extensions, termcolor, tensorflow-io-gcs-filesystem, tensorboard-data-server, six, rsa, pyasn1-modules, protobuf, oauthlib, numpy, MarkupSafe, idna, grpcio, gast, cachetools, absl-py, werkzeug, requests, opt-einsum, keras-preprocessing, importlib-metadata, h5py, google-pasta, google-auth, astunparse, requests-oauthlib, markdown, google-auth-oauthlib, tensorboard, tensorflow\n",
      "Successfully installed MarkupSafe-2.1.2 absl-py-1.4.0 astunparse-1.6.3 cachetools-5.3.0 charset-normalizer-3.0.1 flatbuffers-2.0.7 gast-0.4.0 google-auth-2.16.0 google-auth-oauthlib-0.4.6 google-pasta-0.2.0 grpcio-1.51.1 h5py-3.8.0 idna-3.4 importlib-metadata-6.0.0 keras-2.7.0 keras-preprocessing-1.1.2 libclang-15.0.6.1 markdown-3.4.1 numpy-1.24.1 oauthlib-3.2.2 opt-einsum-3.3.0 protobuf-3.20.3 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.28.2 requests-oauthlib-1.3.1 rsa-4.9 six-1.16.0 tensorboard-2.11.2 tensorboard-data-server-0.6.1 tensorboard-plugin-wit-1.8.1 tensorflow-2.7.0 tensorflow-estimator-2.7.0 tensorflow-io-gcs-filesystem-0.30.0 termcolor-2.2.0 typing-extensions-4.4.0 urllib3-1.26.14 werkzeug-2.2.2 wrapt-1.14.1 zipp-3.12.0\n",
      "Requirement already satisfied: numpy in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (1.24.1)\n",
      "Collecting pandas\n",
      "  Using cached pandas-1.5.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.2 MB)\n",
      "Collecting python-dateutil>=2.8.1\n",
      "  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)\n",
      "Collecting pytz>=2020.1\n",
      "  Using cached pytz-2022.7.1-py2.py3-none-any.whl (499 kB)\n",
      "Requirement already satisfied: numpy>=1.20.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas) (1.24.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)\n",
      "Installing collected packages: pytz, python-dateutil, pandas\n",
      "Successfully installed pandas-1.5.3 python-dateutil-2.8.2 pytz-2022.7.1\n",
      "Collecting scikit-learn\n",
      "  Using cached scikit_learn-1.2.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (9.8 MB)\n",
      "Collecting joblib>=1.1.1\n",
      "  Using cached joblib-1.2.0-py3-none-any.whl (297 kB)\n",
      "Collecting threadpoolctl>=2.0.0\n",
      "  Using cached threadpoolctl-3.1.0-py3-none-any.whl (14 kB)\n",
      "Collecting scipy>=1.3.2\n",
      "  Using cached scipy-1.10.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.5 MB)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from scikit-learn) (1.24.1)\n",
      "Installing collected packages: threadpoolctl, scipy, joblib, scikit-learn\n",
      "Successfully installed joblib-1.2.0 scikit-learn-1.2.1 scipy-1.10.0 threadpoolctl-3.1.0\n",
      "Collecting statsmodels\n",
      "  Using cached statsmodels-0.13.5-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (9.9 MB)\n",
      "Collecting packaging>=21.3\n",
      "  Using cached packaging-23.0-py3-none-any.whl (42 kB)\n",
      "Requirement already satisfied: pandas>=0.25 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from statsmodels) (1.5.3)\n",
      "Collecting patsy>=0.5.2\n",
      "  Using cached patsy-0.5.3-py2.py3-none-any.whl (233 kB)\n",
      "Requirement already satisfied: scipy>=1.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from statsmodels) (1.10.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from statsmodels) (1.24.1)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=0.25->statsmodels) (2022.7.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=0.25->statsmodels) (2.8.2)\n",
      "Requirement already satisfied: six in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from patsy>=0.5.2->statsmodels) (1.16.0)\n",
      "Installing collected packages: patsy, packaging, statsmodels\n",
      "Successfully installed packaging-23.0 patsy-0.5.3 statsmodels-0.13.5\n",
      "Collecting matplotlib\n",
      "  Using cached matplotlib-3.6.3-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (9.4 MB)\n",
      "Collecting pillow>=6.2.0\n",
      "  Using cached Pillow-9.4.0-cp38-cp38-manylinux_2_28_x86_64.whl (3.4 MB)\n",
      "Collecting fonttools>=4.22.0\n",
      "  Using cached fonttools-4.38.0-py3-none-any.whl (965 kB)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib) (23.0)\n",
      "Collecting contourpy>=1.0.1\n",
      "  Using cached contourpy-1.0.7-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (300 kB)\n",
      "Collecting pyparsing>=2.2.1\n",
      "  Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib) (2.8.2)\n",
      "Requirement already satisfied: numpy>=1.19 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib) (1.24.1)\n",
      "Collecting cycler>=0.10\n",
      "  Using cached cycler-0.11.0-py3-none-any.whl (6.4 kB)\n",
      "Collecting kiwisolver>=1.0.1\n",
      "  Using cached kiwisolver-1.4.4-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.2 MB)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
      "Installing collected packages: pyparsing, pillow, kiwisolver, fonttools, cycler, contourpy, matplotlib\n",
      "Successfully installed contourpy-1.0.7 cycler-0.11.0 fonttools-4.38.0 kiwisolver-1.4.4 matplotlib-3.6.3 pillow-9.4.0 pyparsing-3.0.9\n",
      "Collecting future\n",
      "  Using cached future-0.18.3-py3-none-any.whl\n",
      "Installing collected packages: future\n",
      "Successfully installed future-0.18.3\n",
      "Collecting onnx\n",
      "  Using cached onnx-1.13.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.5 MB)\n",
      "Requirement already satisfied: numpy>=1.16.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from onnx) (1.24.1)\n",
      "Requirement already satisfied: typing-extensions>=3.6.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from onnx) (4.4.0)\n",
      "Requirement already satisfied: protobuf<4,>=3.20.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from onnx) (3.20.3)\n",
      "Installing collected packages: onnx\n",
      "Successfully installed onnx-1.13.0\n",
      "Collecting mpi\n",
      "  Using cached mpi-1.0.0-py3-none-any.whl (106 kB)\n",
      "Installing collected packages: mpi\n",
      "Successfully installed mpi-1.0.0\n",
      "Collecting tqdm\n",
      "  Using cached tqdm-4.64.1-py2.py3-none-any.whl (78 kB)\n",
      "Installing collected packages: tqdm\n",
      "Successfully installed tqdm-4.64.1\n",
      "Collecting pydot\n",
      "  Using cached pydot-1.4.2-py2.py3-none-any.whl (21 kB)\n",
      "Requirement already satisfied: pyparsing>=2.1.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pydot) (3.0.9)\n",
      "Installing collected packages: pydot\n",
      "Successfully installed pydot-1.4.2\n",
      "Collecting ipympl\n",
      "  Using cached ipympl-0.9.2-py2.py3-none-any.whl (510 kB)\n",
      "Collecting ipython<9\n",
      "  Using cached ipython-8.9.0-py3-none-any.whl (783 kB)\n",
      "Collecting ipython-genutils\n",
      "  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)\n",
      "Requirement already satisfied: numpy in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipympl) (1.24.1)\n",
      "Requirement already satisfied: matplotlib<4,>=3.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipympl) (3.6.3)\n",
      "Collecting ipywidgets<9,>=7.6.0\n",
      "  Using cached ipywidgets-8.0.4-py3-none-any.whl (137 kB)\n",
      "Requirement already satisfied: pillow in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipympl) (9.4.0)\n",
      "Collecting traitlets<6\n",
      "  Using cached traitlets-5.9.0-py3-none-any.whl (117 kB)\n",
      "Collecting prompt-toolkit<3.1.0,>=3.0.30\n",
      "  Using cached prompt_toolkit-3.0.36-py3-none-any.whl (386 kB)\n",
      "Collecting pexpect>4.3\n",
      "  Using cached pexpect-4.8.0-py2.py3-none-any.whl (59 kB)\n",
      "Collecting backcall\n",
      "  Using cached backcall-0.2.0-py2.py3-none-any.whl (11 kB)\n",
      "Collecting decorator\n",
      "  Using cached decorator-5.1.1-py3-none-any.whl (9.1 kB)\n",
      "Collecting stack-data\n",
      "  Using cached stack_data-0.6.2-py3-none-any.whl (24 kB)\n",
      "Collecting pygments>=2.4.0\n",
      "  Using cached Pygments-2.14.0-py3-none-any.whl (1.1 MB)\n",
      "Collecting pickleshare\n",
      "  Using cached pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)\n",
      "Collecting jedi>=0.16\n",
      "  Using cached jedi-0.18.2-py2.py3-none-any.whl (1.6 MB)\n",
      "Collecting matplotlib-inline\n",
      "  Using cached matplotlib_inline-0.1.6-py3-none-any.whl (9.4 kB)\n",
      "Collecting ipykernel>=4.5.1\n",
      "  Using cached ipykernel-6.21.1-py3-none-any.whl (149 kB)\n",
      "Collecting widgetsnbextension~=4.0\n",
      "  Using cached widgetsnbextension-4.0.5-py3-none-any.whl (2.0 MB)\n",
      "Collecting jupyterlab-widgets~=3.0\n",
      "  Using cached jupyterlab_widgets-3.0.5-py3-none-any.whl (384 kB)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (0.11.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (1.4.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (2.8.2)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (4.38.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (23.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (3.0.9)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib<4,>=3.4.0->ipympl) (1.0.7)\n",
      "Collecting nest-asyncio\n",
      "  Using cached nest_asyncio-1.5.6-py3-none-any.whl (5.2 kB)\n",
      "Collecting psutil\n",
      "  Using cached psutil-5.9.4-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (280 kB)\n",
      "Collecting pyzmq>=17\n",
      "  Using cached pyzmq-25.0.0-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.1 MB)\n",
      "Collecting debugpy>=1.6.5\n",
      "  Using cached debugpy-1.6.6-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "Collecting tornado>=6.1\n",
      "  Using cached tornado-6.2-cp37-abi3-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (423 kB)\n",
      "Collecting jupyter-core!=5.0.*,>=4.12\n",
      "  Using cached jupyter_core-5.2.0-py3-none-any.whl (94 kB)\n",
      "Collecting comm>=0.1.1\n",
      "  Using cached comm-0.1.2-py3-none-any.whl (6.5 kB)\n",
      "Collecting jupyter-client>=6.1.12\n",
      "  Using cached jupyter_client-8.0.2-py3-none-any.whl (103 kB)\n",
      "Collecting parso<0.9.0,>=0.8.0\n",
      "  Using cached parso-0.8.3-py2.py3-none-any.whl (100 kB)\n",
      "Collecting ptyprocess>=0.5\n",
      "  Using cached ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)\n",
      "Collecting wcwidth\n",
      "  Using cached wcwidth-0.2.6-py2.py3-none-any.whl (29 kB)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib<4,>=3.4.0->ipympl) (1.16.0)\n",
      "Collecting executing>=1.2.0\n",
      "  Using cached executing-1.2.0-py2.py3-none-any.whl (24 kB)\n",
      "Collecting pure-eval\n",
      "  Using cached pure_eval-0.2.2-py3-none-any.whl (11 kB)\n",
      "Collecting asttokens>=2.1.0\n",
      "  Using cached asttokens-2.2.1-py2.py3-none-any.whl (26 kB)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets<9,>=7.6.0->ipympl) (6.0.0)\n",
      "Collecting platformdirs>=2.5\n",
      "  Using cached platformdirs-2.6.2-py3-none-any.whl (14 kB)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=6.1.12->ipykernel>=4.5.1->ipywidgets<9,>=7.6.0->ipympl) (3.12.0)\n",
      "Installing collected packages: wcwidth, pure-eval, ptyprocess, pickleshare, ipython-genutils, executing, backcall, widgetsnbextension, traitlets, tornado, pyzmq, pygments, psutil, prompt-toolkit, platformdirs, pexpect, parso, nest-asyncio, jupyterlab-widgets, decorator, debugpy, asttokens, stack-data, matplotlib-inline, jupyter-core, jedi, comm, jupyter-client, ipython, ipykernel, ipywidgets, ipympl\n",
      "Successfully installed asttokens-2.2.1 backcall-0.2.0 comm-0.1.2 debugpy-1.6.6 decorator-5.1.1 executing-1.2.0 ipykernel-6.21.1 ipympl-0.9.2 ipython-8.9.0 ipython-genutils-0.2.0 ipywidgets-8.0.4 jedi-0.18.2 jupyter-client-8.0.2 jupyter-core-5.2.0 jupyterlab-widgets-3.0.5 matplotlib-inline-0.1.6 nest-asyncio-1.5.6 parso-0.8.3 pexpect-4.8.0 pickleshare-0.7.5 platformdirs-2.6.2 prompt-toolkit-3.0.36 psutil-5.9.4 ptyprocess-0.7.0 pure-eval-0.2.2 pygments-2.14.0 pyzmq-25.0.0 stack-data-0.6.2 tornado-6.2 traitlets-5.9.0 wcwidth-0.2.6 widgetsnbextension-4.0.5\n",
      "Collecting seaborn\n",
      "  Using cached seaborn-0.12.2-py3-none-any.whl (293 kB)\n",
      "Requirement already satisfied: numpy!=1.24.0,>=1.17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from seaborn) (1.24.1)\n",
      "Requirement already satisfied: matplotlib!=3.6.1,>=3.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from seaborn) (3.6.3)\n",
      "Requirement already satisfied: pandas>=0.25 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from seaborn) (1.5.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (2.8.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (3.0.9)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (23.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (9.4.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (4.38.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (1.4.4)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (1.0.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib!=3.6.1,>=3.1->seaborn) (0.11.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=0.25->seaborn) (2022.7.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib!=3.6.1,>=3.1->seaborn) (1.16.0)\n",
      "Installing collected packages: seaborn\n",
      "Successfully installed seaborn-0.12.2\n",
      "Collecting tabulate\n",
      "  Using cached tabulate-0.9.0-py3-none-any.whl (35 kB)\n",
      "Installing collected packages: tabulate\n",
      "Successfully installed tabulate-0.9.0\n",
      "Collecting xgboost\n",
      "  Using cached xgboost-1.7.3-py3-none-manylinux2014_x86_64.whl (193.6 MB)\n",
      "Requirement already satisfied: scipy in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from xgboost) (1.10.0)\n",
      "Requirement already satisfied: numpy in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from xgboost) (1.24.1)\n",
      "Installing collected packages: xgboost\n",
      "Successfully installed xgboost-1.7.3\n",
      "Collecting catboost\n",
      "  Using cached catboost-1.1.1-cp38-none-manylinux1_x86_64.whl (76.6 MB)\n",
      "Requirement already satisfied: six in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from catboost) (1.16.0)\n",
      "Requirement already satisfied: scipy in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from catboost) (1.10.0)\n",
      "Requirement already satisfied: pandas>=0.24.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from catboost) (1.5.3)\n",
      "Collecting plotly\n",
      "  Using cached plotly-5.13.0-py2.py3-none-any.whl (15.2 MB)\n",
      "Collecting graphviz\n",
      "  Using cached graphviz-0.20.1-py3-none-any.whl (47 kB)\n",
      "Requirement already satisfied: matplotlib in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from catboost) (3.6.3)\n",
      "Requirement already satisfied: numpy>=1.16.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from catboost) (1.24.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=0.24.0->catboost) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=0.24.0->catboost) (2022.7.1)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (23.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (1.4.4)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (4.38.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (9.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (3.0.9)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib->catboost) (1.0.7)\n",
      "Collecting tenacity>=6.2.0\n",
      "  Using cached tenacity-8.1.0-py3-none-any.whl (23 kB)\n",
      "Installing collected packages: tenacity, graphviz, plotly, catboost\n",
      "Successfully installed catboost-1.1.1 graphviz-0.20.1 plotly-5.13.0 tenacity-8.1.0\n",
      "Collecting bokeh\n",
      "  Using cached bokeh-3.0.3-py3-none-any.whl (16.5 MB)\n",
      "Requirement already satisfied: contourpy>=1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (1.0.7)\n",
      "Requirement already satisfied: pillow>=7.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (9.4.0)\n",
      "Requirement already satisfied: numpy>=1.11.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (1.24.1)\n",
      "Collecting PyYAML>=3.10\n",
      "  Using cached PyYAML-6.0-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (701 kB)\n",
      "Collecting Jinja2>=2.9\n",
      "  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)\n",
      "Requirement already satisfied: tornado>=5.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (6.2)\n",
      "Collecting xyzservices>=2021.09.1\n",
      "  Using cached xyzservices-2022.9.0-py3-none-any.whl (55 kB)\n",
      "Requirement already satisfied: packaging>=16.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (23.0)\n",
      "Requirement already satisfied: pandas>=1.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bokeh) (1.5.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from Jinja2>=2.9->bokeh) (2.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=1.2->bokeh) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=1.2->bokeh) (2022.7.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.8.1->pandas>=1.2->bokeh) (1.16.0)\n",
      "Installing collected packages: xyzservices, PyYAML, Jinja2, bokeh\n",
      "Successfully installed Jinja2-3.1.2 PyYAML-6.0 bokeh-3.0.3 xyzservices-2022.9.0\n",
      "Cloning into 'mlperf-logging'...\n",
      "Obtaining file:///tmp/mlperf-logging\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: pandas>=1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from mlperf-logging==2.1.0) (1.5.3)\n",
      "Requirement already satisfied: pyyaml>=5.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from mlperf-logging==2.1.0) (6.0)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from mlperf-logging==2.1.0) (1.24.1)\n",
      "Requirement already satisfied: scipy>=1.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from mlperf-logging==2.1.0) (1.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=1.0->mlperf-logging==2.1.0) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pandas>=1.0->mlperf-logging==2.1.0) (2022.7.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.8.1->pandas>=1.0->mlperf-logging==2.1.0) (1.16.0)\n",
      "Installing collected packages: mlperf-logging\n",
      "  Running setup.py develop for mlperf-logging\n",
      "Successfully installed mlperf-logging-2.1.0\n",
      "Requirement already satisfied: tensorboard in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (2.11.2)\n",
      "Requirement already satisfied: grpcio>=1.24.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (1.51.1)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (3.4.1)\n",
      "Requirement already satisfied: wheel>=0.26 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (0.37.1)\n",
      "Requirement already satisfied: protobuf<4,>=3.9.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (3.20.3)\n",
      "Requirement already satisfied: numpy>=1.12.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (1.24.1)\n",
      "Requirement already satisfied: google-auth<3,>=1.6.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (2.16.0)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (2.28.2)\n",
      "Requirement already satisfied: absl-py>=0.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (1.4.0)\n",
      "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (0.6.1)\n",
      "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (0.4.6)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (2.2.2)\n",
      "Requirement already satisfied: setuptools>=41.0.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (65.6.3)\n",
      "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from tensorboard) (1.8.1)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard) (4.9)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard) (5.3.0)\n",
      "Requirement already satisfied: six>=1.9.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard) (1.16.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard) (0.2.8)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard) (1.3.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from markdown>=2.6.8->tensorboard) (6.0.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard) (1.26.14)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard) (2022.12.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard) (3.0.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from werkzeug>=1.0.1->tensorboard) (2.1.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard) (3.12.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard) (3.2.2)\n",
      "Collecting jupyter\n",
      "  Using cached jupyter-1.0.0-py2.py3-none-any.whl (2.7 kB)\n",
      "Requirement already satisfied: ipywidgets in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter) (8.0.4)\n",
      "Collecting notebook\n",
      "  Using cached notebook-6.5.2-py3-none-any.whl (439 kB)\n",
      "Collecting nbconvert\n",
      "  Using cached nbconvert-7.2.9-py3-none-any.whl (274 kB)\n",
      "Collecting qtconsole\n",
      "  Using cached qtconsole-5.4.0-py3-none-any.whl (121 kB)\n",
      "Collecting jupyter-console\n",
      "  Using cached jupyter_console-6.4.4-py3-none-any.whl (22 kB)\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter) (6.21.1)\n",
      "Requirement already satisfied: nest-asyncio in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (1.5.6)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (8.0.2)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (5.2.0)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (0.1.6)\n",
      "Requirement already satisfied: ipython>=7.23.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (8.9.0)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (6.2)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (5.9.0)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (25.0.0)\n",
      "Requirement already satisfied: packaging in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (23.0)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (1.6.6)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (5.9.4)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->jupyter) (0.1.2)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipywidgets->jupyter) (4.0.5)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipywidgets->jupyter) (3.0.5)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-console->jupyter) (3.0.36)\n",
      "Requirement already satisfied: pygments in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-console->jupyter) (2.14.0)\n",
      "Collecting mistune<3,>=2.0.3\n",
      "  Using cached mistune-2.0.4-py2.py3-none-any.whl (24 kB)\n",
      "Collecting pandocfilters>=1.4.1\n",
      "  Using cached pandocfilters-1.5.0-py2.py3-none-any.whl (8.7 kB)\n",
      "Requirement already satisfied: markupsafe>=2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert->jupyter) (2.1.2)\n",
      "Collecting bleach\n",
      "  Using cached bleach-6.0.0-py3-none-any.whl (162 kB)\n",
      "Collecting defusedxml\n",
      "  Using cached defusedxml-0.7.1-py2.py3-none-any.whl (25 kB)\n",
      "Collecting jupyterlab-pygments\n",
      "  Using cached jupyterlab_pygments-0.2.2-py2.py3-none-any.whl (21 kB)\n",
      "Requirement already satisfied: jinja2>=3.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert->jupyter) (3.1.2)\n",
      "Requirement already satisfied: importlib-metadata>=3.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert->jupyter) (6.0.0)\n",
      "Collecting beautifulsoup4\n",
      "  Using cached beautifulsoup4-4.11.2-py3-none-any.whl (129 kB)\n",
      "Collecting nbformat>=5.1\n",
      "  Using cached nbformat-5.7.3-py3-none-any.whl (78 kB)\n",
      "Collecting nbclient>=0.5.0\n",
      "  Using cached nbclient-0.7.2-py3-none-any.whl (71 kB)\n",
      "Collecting tinycss2\n",
      "  Using cached tinycss2-1.2.1-py3-none-any.whl (21 kB)\n",
      "Collecting Send2Trash>=1.8.0\n",
      "  Using cached Send2Trash-1.8.0-py3-none-any.whl (18 kB)\n",
      "Collecting terminado>=0.8.3\n",
      "  Using cached terminado-0.17.1-py3-none-any.whl (17 kB)\n",
      "Collecting nbclassic>=0.4.7\n",
      "  Using cached nbclassic-0.5.1-py3-none-any.whl (10.0 MB)\n",
      "Collecting prometheus-client\n",
      "  Using cached prometheus_client-0.16.0-py3-none-any.whl (122 kB)\n",
      "Requirement already satisfied: ipython-genutils in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook->jupyter) (0.2.0)\n",
      "Collecting argon2-cffi\n",
      "  Using cached argon2_cffi-21.3.0-py3-none-any.whl (14 kB)\n",
      "Collecting qtpy>=2.0.1\n",
      "  Using cached QtPy-2.3.0-py3-none-any.whl (83 kB)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=3.6->nbconvert->jupyter) (3.12.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (4.8.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.18.2)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.7.5)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.2.0)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.6.2)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel->jupyter) (5.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel->jupyter) (2.8.2)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->jupyter) (2.6.2)\n",
      "Collecting jupyter-server>=1.17.0\n",
      "  Using cached jupyter_server-2.2.1-py3-none-any.whl (365 kB)\n",
      "Collecting notebook-shim>=0.1.0\n",
      "  Using cached notebook_shim-0.2.2-py3-none-any.whl (13 kB)\n",
      "Collecting fastjsonschema\n",
      "  Using cached fastjsonschema-2.16.2-py3-none-any.whl (22 kB)\n",
      "Collecting jsonschema>=2.6\n",
      "  Using cached jsonschema-4.17.3-py3-none-any.whl (90 kB)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->jupyter-console->jupyter) (0.2.6)\n",
      "Requirement already satisfied: ptyprocess in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from terminado>=0.8.3->notebook->jupyter) (0.7.0)\n",
      "Collecting argon2-cffi-bindings\n",
      "  Using cached argon2_cffi_bindings-21.2.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (86 kB)\n",
      "Collecting soupsieve>1.2\n",
      "  Using cached soupsieve-2.3.2.post1-py3-none-any.whl (37 kB)\n",
      "Requirement already satisfied: six>=1.9.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bleach->nbconvert->jupyter) (1.16.0)\n",
      "Collecting webencodings\n",
      "  Using cached webencodings-0.5.1-py2.py3-none-any.whl (11 kB)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel->jupyter) (0.8.3)\n",
      "Collecting importlib-resources>=1.4.0\n",
      "  Using cached importlib_resources-5.10.2-py3-none-any.whl (34 kB)\n",
      "Collecting pkgutil-resolve-name>=1.3.10\n",
      "  Using cached pkgutil_resolve_name-1.3.10-py3-none-any.whl (4.7 kB)\n",
      "Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0\n",
      "  Using cached pyrsistent-0.19.3-py3-none-any.whl (57 kB)\n",
      "Collecting attrs>=17.4.0\n",
      "  Using cached attrs-22.2.0-py3-none-any.whl (60 kB)\n",
      "Collecting jupyter-events>=0.4.0\n",
      "  Using cached jupyter_events-0.6.3-py3-none-any.whl (18 kB)\n",
      "Collecting anyio>=3.1.0\n",
      "  Using cached anyio-3.6.2-py3-none-any.whl (80 kB)\n",
      "Collecting websocket-client\n",
      "  Using cached websocket_client-1.5.1-py3-none-any.whl (55 kB)\n",
      "Collecting jupyter-server-terminals\n",
      "  Using cached jupyter_server_terminals-0.4.4-py3-none-any.whl (13 kB)\n",
      "Collecting cffi>=1.0.1\n",
      "  Using cached cffi-1.15.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (442 kB)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (1.2.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (2.2.1)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (0.2.2)\n",
      "Collecting sniffio>=1.1\n",
      "  Using cached sniffio-1.3.0-py3-none-any.whl (10 kB)\n",
      "Requirement already satisfied: idna>=2.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook->jupyter) (3.4)\n",
      "Collecting pycparser\n",
      "  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)\n",
      "Collecting rfc3986-validator>=0.1.1\n",
      "  Using cached rfc3986_validator-0.1.1-py2.py3-none-any.whl (4.2 kB)\n",
      "Collecting python-json-logger>=2.0.4\n",
      "  Using cached python_json_logger-2.0.4-py3-none-any.whl (7.8 kB)\n",
      "Collecting rfc3339-validator\n",
      "  Using cached rfc3339_validator-0.1.4-py2.py3-none-any.whl (3.5 kB)\n",
      "Requirement already satisfied: pyyaml>=5.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook->jupyter) (6.0)\n",
      "Collecting webcolors>=1.11\n",
      "  Using cached webcolors-1.12-py3-none-any.whl (9.9 kB)\n",
      "Collecting jsonpointer>1.13\n",
      "  Using cached jsonpointer-2.3-py2.py3-none-any.whl (7.8 kB)\n",
      "Collecting isoduration\n",
      "  Using cached isoduration-20.11.0-py3-none-any.whl (11 kB)\n",
      "Collecting uri-template\n",
      "  Using cached uri_template-1.2.0-py3-none-any.whl (10 kB)\n",
      "Collecting fqdn\n",
      "  Using cached fqdn-1.5.1-py3-none-any.whl (9.1 kB)\n",
      "Collecting arrow>=0.15.0\n",
      "  Using cached arrow-1.2.3-py3-none-any.whl (66 kB)\n",
      "Installing collected packages: webencodings, Send2Trash, mistune, fastjsonschema, websocket-client, webcolors, uri-template, tinycss2, terminado, soupsieve, sniffio, rfc3986-validator, rfc3339-validator, qtpy, python-json-logger, pyrsistent, pycparser, prometheus-client, pkgutil-resolve-name, pandocfilters, jupyterlab-pygments, jsonpointer, importlib-resources, fqdn, defusedxml, bleach, attrs, jupyter-server-terminals, jsonschema, cffi, beautifulsoup4, arrow, anyio, nbformat, isoduration, argon2-cffi-bindings, nbclient, argon2-cffi, qtconsole, nbconvert, jupyter-events, jupyter-console, jupyter-server, notebook-shim, nbclassic, notebook, jupyter\n",
      "Successfully installed Send2Trash-1.8.0 anyio-3.6.2 argon2-cffi-21.3.0 argon2-cffi-bindings-21.2.0 arrow-1.2.3 attrs-22.2.0 beautifulsoup4-4.11.2 bleach-6.0.0 cffi-1.15.1 defusedxml-0.7.1 fastjsonschema-2.16.2 fqdn-1.5.1 importlib-resources-5.10.2 isoduration-20.11.0 jsonpointer-2.3 jsonschema-4.17.3 jupyter-1.0.0 jupyter-console-6.4.4 jupyter-events-0.6.3 jupyter-server-2.2.1 jupyter-server-terminals-0.4.4 jupyterlab-pygments-0.2.2 mistune-2.0.4 nbclassic-0.5.1 nbclient-0.7.2 nbconvert-7.2.9 nbformat-5.7.3 notebook-6.5.2 notebook-shim-0.2.2 pandocfilters-1.5.0 pkgutil-resolve-name-1.3.10 prometheus-client-0.16.0 pycparser-2.21 pyrsistent-0.19.3 python-json-logger-2.0.4 qtconsole-5.4.0 qtpy-2.3.0 rfc3339-validator-0.1.4 rfc3986-validator-0.1.1 sniffio-1.3.0 soupsieve-2.3.2.post1 terminado-0.17.1 tinycss2-1.2.1 uri-template-1.2.0 webcolors-1.12 webencodings-0.5.1 websocket-client-1.5.1\n",
      "Collecting jupyterlab\n",
      "  Using cached jupyterlab-3.6.1-py3-none-any.whl (8.9 MB)\n",
      "Requirement already satisfied: packaging in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (23.0)\n",
      "Requirement already satisfied: nbclassic in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (0.5.1)\n",
      "Requirement already satisfied: tornado>=6.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (6.2)\n",
      "Collecting jupyterlab-server~=2.19\n",
      "  Using cached jupyterlab_server-2.19.0-py3-none-any.whl (56 kB)\n",
      "Collecting jupyter-server-ydoc<0.7.0,>=0.6.0\n",
      "  Using cached jupyter_server_ydoc-0.6.1-py3-none-any.whl (11 kB)\n",
      "Collecting jupyter-ydoc~=0.2.2\n",
      "  Using cached jupyter_ydoc-0.2.2-py3-none-any.whl (5.6 kB)\n",
      "Requirement already satisfied: jupyter-server<3,>=1.16.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (2.2.1)\n",
      "Requirement already satisfied: notebook<7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (6.5.2)\n",
      "Collecting tomli\n",
      "  Using cached tomli-2.0.1-py3-none-any.whl (12 kB)\n",
      "Requirement already satisfied: jupyter-core in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (5.2.0)\n",
      "Requirement already satisfied: jinja2>=2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (3.1.2)\n",
      "Requirement already satisfied: ipython in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab) (8.9.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jinja2>=2.1->jupyterlab) (2.1.2)\n",
      "Requirement already satisfied: argon2-cffi in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (21.3.0)\n",
      "Requirement already satisfied: nbconvert>=6.4.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (7.2.9)\n",
      "Requirement already satisfied: jupyter-client>=7.4.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (8.0.2)\n",
      "Requirement already satisfied: jupyter-server-terminals in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (0.4.4)\n",
      "Requirement already satisfied: nbformat>=5.3.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (5.7.3)\n",
      "Requirement already satisfied: jupyter-events>=0.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (0.6.3)\n",
      "Requirement already satisfied: send2trash in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (1.8.0)\n",
      "Requirement already satisfied: anyio>=3.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (3.6.2)\n",
      "Requirement already satisfied: terminado>=0.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (0.17.1)\n",
      "Requirement already satisfied: pyzmq>=24 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (25.0.0)\n",
      "Requirement already satisfied: traitlets>=5.6.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (5.9.0)\n",
      "Requirement already satisfied: websocket-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (1.5.1)\n",
      "Requirement already satisfied: prometheus-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server<3,>=1.16.0->jupyterlab) (0.16.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core->jupyterlab) (2.6.2)\n",
      "Collecting ypy-websocket<0.9.0,>=0.8.2\n",
      "  Using cached ypy_websocket-0.8.2-py3-none-any.whl (10 kB)\n",
      "Collecting jupyter-server-fileid<1,>=0.6.0\n",
      "  Using cached jupyter_server_fileid-0.6.0-py3-none-any.whl (15 kB)\n",
      "Requirement already satisfied: importlib-metadata>=3.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-ydoc~=0.2.2->jupyterlab) (6.0.0)\n",
      "Collecting y-py<0.6.0,>=0.5.3\n",
      "  Using cached y_py-0.5.5-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.6 MB)\n",
      "Requirement already satisfied: jsonschema>=4.17.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab-server~=2.19->jupyterlab) (4.17.3)\n",
      "Requirement already satisfied: requests>=2.28 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterlab-server~=2.19->jupyterlab) (2.28.2)\n",
      "Collecting babel>=2.10\n",
      "  Using cached Babel-2.11.0-py3-none-any.whl (9.5 MB)\n",
      "Collecting json5>=0.9.0\n",
      "  Using cached json5-0.9.11-py2.py3-none-any.whl (19 kB)\n",
      "Requirement already satisfied: nest-asyncio>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook<7->jupyterlab) (1.5.6)\n",
      "Requirement already satisfied: ipython-genutils in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook<7->jupyterlab) (0.2.0)\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook<7->jupyterlab) (6.21.1)\n",
      "Requirement already satisfied: notebook-shim>=0.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbclassic->jupyterlab) (0.2.2)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (2.14.0)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (0.7.5)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (0.2.0)\n",
      "Requirement already satisfied: matplotlib-inline in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (0.1.6)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (0.18.2)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.30 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (3.0.36)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (0.6.2)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (5.1.1)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython->jupyterlab) (4.8.0)\n",
      "Requirement already satisfied: idna>=2.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab) (3.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab) (1.3.0)\n",
      "Requirement already satisfied: pytz>=2015.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from babel>=2.10->jupyterlab-server~=2.19->jupyterlab) (2022.7.1)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=3.6->jupyter-ydoc~=0.2.2->jupyterlab) (3.12.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython->jupyterlab) (0.8.3)\n",
      "Requirement already satisfied: importlib-resources>=1.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (5.10.2)\n",
      "Requirement already satisfied: pkgutil-resolve-name>=1.3.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (1.3.10)\n",
      "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (0.19.3)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (22.2.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=7.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (2.8.2)\n",
      "Requirement already satisfied: python-json-logger>=2.0.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server<3,>=1.16.0->jupyterlab) (2.0.4)\n",
      "Requirement already satisfied: pyyaml>=5.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server<3,>=1.16.0->jupyterlab) (6.0)\n",
      "Requirement already satisfied: rfc3339-validator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server<3,>=1.16.0->jupyterlab) (0.1.4)\n",
      "Requirement already satisfied: rfc3986-validator>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server<3,>=1.16.0->jupyterlab) (0.1.1)\n",
      "Collecting jupyter-events>=0.4.0\n",
      "  Using cached jupyter_events-0.5.0-py3-none-any.whl (17 kB)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (4.11.2)\n",
      "Requirement already satisfied: bleach in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (6.0.0)\n",
      "Requirement already satisfied: jupyterlab-pygments in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (0.2.2)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (1.5.0)\n",
      "Requirement already satisfied: mistune<3,>=2.0.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (2.0.4)\n",
      "Requirement already satisfied: defusedxml in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (0.7.1)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (0.7.2)\n",
      "Requirement already satisfied: tinycss2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (1.2.1)\n",
      "Requirement already satisfied: fastjsonschema in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbformat>=5.3.0->jupyter-server<3,>=1.16.0->jupyterlab) (2.16.2)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pexpect>4.3->ipython->jupyterlab) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit<3.1.0,>=3.0.30->ipython->jupyterlab) (0.2.6)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests>=2.28->jupyterlab-server~=2.19->jupyterlab) (2022.12.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests>=2.28->jupyterlab-server~=2.19->jupyterlab) (3.0.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from requests>=2.28->jupyterlab-server~=2.19->jupyterlab) (1.26.14)\n",
      "Collecting aiosqlite<1,>=0.17.0\n",
      "  Using cached aiosqlite-0.18.0-py3-none-any.whl (15 kB)\n",
      "Collecting aiofiles<23,>=22.1.0\n",
      "  Using cached aiofiles-22.1.0-py3-none-any.whl (14 kB)\n",
      "Requirement already satisfied: argon2-cffi-bindings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi->jupyter-server<3,>=1.16.0->jupyterlab) (21.2.0)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook<7->jupyterlab) (1.6.6)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook<7->jupyterlab) (0.1.2)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook<7->jupyterlab) (5.9.4)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython->jupyterlab) (0.2.2)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython->jupyterlab) (2.2.1)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython->jupyterlab) (1.2.0)\n",
      "Requirement already satisfied: six in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from asttokens>=2.1.0->stack-data->ipython->jupyterlab) (1.16.0)\n",
      "Requirement already satisfied: isoduration in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (20.11.0)\n",
      "Requirement already satisfied: uri-template in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (1.2.0)\n",
      "Requirement already satisfied: jsonpointer>1.13 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (2.3)\n",
      "Requirement already satisfied: fqdn in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (1.5.1)\n",
      "Requirement already satisfied: webcolors>=1.11 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (1.12)\n",
      "Requirement already satisfied: cffi>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi-bindings->argon2-cffi->jupyter-server<3,>=1.16.0->jupyterlab) (1.15.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from beautifulsoup4->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (2.3.2.post1)\n",
      "Requirement already satisfied: webencodings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bleach->nbconvert>=6.4.4->jupyter-server<3,>=1.16.0->jupyterlab) (0.5.1)\n",
      "Requirement already satisfied: pycparser in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->jupyter-server<3,>=1.16.0->jupyterlab) (2.21)\n",
      "Requirement already satisfied: arrow>=0.15.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from isoduration->jsonschema>=4.17.3->jupyterlab-server~=2.19->jupyterlab) (1.2.3)\n",
      "Installing collected packages: y-py, json5, tomli, babel, aiosqlite, aiofiles, ypy-websocket, jupyter-ydoc, jupyter-events, jupyterlab-server, jupyter-server-fileid, jupyter-server-ydoc, jupyterlab\n",
      "  Attempting uninstall: jupyter-events\n",
      "    Found existing installation: jupyter-events 0.6.3\n",
      "    Uninstalling jupyter-events-0.6.3:\n",
      "      Successfully uninstalled jupyter-events-0.6.3\n",
      "Successfully installed aiofiles-22.1.0 aiosqlite-0.18.0 babel-2.11.0 json5-0.9.11 jupyter-events-0.5.0 jupyter-server-fileid-0.6.0 jupyter-server-ydoc-0.6.1 jupyter-ydoc-0.2.2 jupyterlab-3.6.1 jupyterlab-server-2.19.0 tomli-2.0.1 y-py-0.5.5 ypy-websocket-0.8.2\n",
      "Collecting jupyterthemes\n",
      "  Using cached jupyterthemes-0.20.0-py2.py3-none-any.whl (7.0 MB)\n",
      "Requirement already satisfied: jupyter-core in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (5.2.0)\n",
      "Requirement already satisfied: ipython>=5.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (8.9.0)\n",
      "Collecting lesscpy>=0.11.2\n",
      "  Using cached lesscpy-0.15.1-py2.py3-none-any.whl (46 kB)\n",
      "Requirement already satisfied: matplotlib>=1.4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (3.6.3)\n",
      "Requirement already satisfied: notebook>=5.6.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (6.5.2)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (5.1.1)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.7.5)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.2.0)\n",
      "Requirement already satisfied: traitlets>=5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (5.9.0)\n",
      "Requirement already satisfied: matplotlib-inline in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.1.6)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.6.2)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.18.2)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (4.8.0)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.30 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (3.0.36)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (2.14.0)\n",
      "Collecting ply\n",
      "  Using cached ply-3.11-py2.py3-none-any.whl (49 kB)\n",
      "Requirement already satisfied: numpy>=1.19 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.24.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (4.38.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (3.0.9)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (2.8.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (0.11.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (9.4.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.0.7)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.4.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (23.0)\n",
      "Requirement already satisfied: jupyter-client>=5.3.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (8.0.2)\n",
      "Requirement already satisfied: nbformat in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (5.7.3)\n",
      "Requirement already satisfied: terminado>=0.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.17.1)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (6.2)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (25.0.0)\n",
      "Requirement already satisfied: nest-asyncio>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (1.5.6)\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (6.21.1)\n",
      "Requirement already satisfied: nbclassic>=0.4.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.5.1)\n",
      "Requirement already satisfied: argon2-cffi in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (21.3.0)\n",
      "Requirement already satisfied: jinja2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (3.1.2)\n",
      "Requirement already satisfied: nbconvert>=5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (7.2.9)\n",
      "Requirement already satisfied: prometheus-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.16.0)\n",
      "Requirement already satisfied: Send2Trash>=1.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (1.8.0)\n",
      "Requirement already satisfied: ipython-genutils in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.2.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core->jupyterthemes) (2.6.2)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython>=5.4.1->jupyterthemes) (0.8.3)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=5.3.4->notebook>=5.6.0->jupyterthemes) (6.0.0)\n",
      "Requirement already satisfied: jupyter-server>=1.17.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (2.2.1)\n",
      "Requirement already satisfied: notebook-shim>=0.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: jupyterlab-pygments in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: defusedxml in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.7.1)\n",
      "Requirement already satisfied: tinycss2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (1.2.1)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (1.5.0)\n",
      "Requirement already satisfied: markupsafe>=2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.1.2)\n",
      "Requirement already satisfied: mistune<3,>=2.0.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.0.4)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (4.11.2)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.7.2)\n",
      "Requirement already satisfied: bleach in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (6.0.0)\n",
      "Requirement already satisfied: fastjsonschema in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbformat->notebook>=5.6.0->jupyterthemes) (2.16.2)\n",
      "Requirement already satisfied: jsonschema>=2.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbformat->notebook>=5.6.0->jupyterthemes) (4.17.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pexpect>4.3->ipython>=5.4.1->jupyterthemes) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit<3.1.0,>=3.0.30->ipython>=5.4.1->jupyterthemes) (0.2.6)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib>=1.4.3->jupyterthemes) (1.16.0)\n",
      "Requirement already satisfied: argon2-cffi-bindings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi->notebook>=5.6.0->jupyterthemes) (21.2.0)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (0.1.2)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (5.9.4)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (1.6.6)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (2.2.1)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (1.2.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=5.3.4->notebook>=5.6.0->jupyterthemes) (3.12.0)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (22.2.0)\n",
      "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.19.3)\n",
      "Requirement already satisfied: importlib-resources>=1.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (5.10.2)\n",
      "Requirement already satisfied: pkgutil-resolve-name>=1.3.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.3.10)\n",
      "Requirement already satisfied: websocket-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (1.5.1)\n",
      "Requirement already satisfied: jupyter-server-terminals in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.4.4)\n",
      "Requirement already satisfied: anyio>=3.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (3.6.2)\n",
      "Requirement already satisfied: jupyter-events>=0.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.5.0)\n",
      "Requirement already satisfied: cffi>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=5.6.0->jupyterthemes) (1.15.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from beautifulsoup4->nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.3.2.post1)\n",
      "Requirement already satisfied: webencodings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bleach->nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.5.1)\n",
      "Requirement already satisfied: idna>=2.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (3.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (1.3.0)\n",
      "Requirement already satisfied: pycparser in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=5.6.0->jupyterthemes) (2.21)\n",
      "Requirement already satisfied: pyyaml in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (6.0)\n",
      "Requirement already satisfied: python-json-logger in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server>=1.17.0->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (2.0.4)\n",
      "Requirement already satisfied: jsonpointer>1.13 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (2.3)\n",
      "Requirement already satisfied: rfc3986-validator>0.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.1.1)\n",
      "Requirement already satisfied: isoduration in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (20.11.0)\n",
      "Requirement already satisfied: webcolors>=1.11 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.12)\n",
      "Requirement already satisfied: uri-template in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.2.0)\n",
      "Requirement already satisfied: fqdn in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.5.1)\n",
      "Requirement already satisfied: rfc3339-validator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.1.4)\n",
      "Requirement already satisfied: arrow>=0.15.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from isoduration->jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.2.3)\n",
      "Installing collected packages: ply, lesscpy, jupyterthemes\n",
      "Successfully installed jupyterthemes-0.20.0 lesscpy-0.15.1 ply-3.11\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (6.21.1)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (0.1.6)\n",
      "Requirement already satisfied: packaging in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (23.0)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (1.6.6)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (25.0.0)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (6.2)\n",
      "Requirement already satisfied: ipython>=7.23.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (8.9.0)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.9.0)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.9.4)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (8.0.2)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.2.0)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (0.1.2)\n",
      "Requirement already satisfied: nest-asyncio in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (1.5.6)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (5.1.1)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.2.0)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.30 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (3.0.36)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (2.14.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (4.8.0)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.6.2)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.7.5)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.18.2)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel) (6.0.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.8.2)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (2.6.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=6.1.12->ipykernel) (3.12.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pexpect>4.3->ipython>=7.23.1->ipykernel) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit<3.1.0,>=3.0.30->ipython>=7.23.1->ipykernel) (0.2.6)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.16.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (1.2.0)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (0.2.2)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.2.1)\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (6.21.1)\n",
      "Requirement already satisfied: ipython>=7.23.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (8.9.0)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.9.0)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.2.0)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (1.6.6)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (0.1.2)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (6.2)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (5.9.4)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (25.0.0)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (0.1.6)\n",
      "Requirement already satisfied: nest-asyncio in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (1.5.6)\n",
      "Requirement already satisfied: packaging in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (23.0)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel) (8.0.2)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (4.8.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.18.2)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.30 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (3.0.36)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (5.1.1)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (2.14.0)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.2.0)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.6.2)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=7.23.1->ipykernel) (0.7.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.8.2)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=6.1.12->ipykernel) (6.0.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (2.6.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=6.1.12->ipykernel) (3.12.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pexpect>4.3->ipython>=7.23.1->ipykernel) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit<3.1.0,>=3.0.30->ipython>=7.23.1->ipykernel) (0.2.6)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.16.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.2.1)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (0.2.2)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (1.2.0)\n",
      "Installed kernelspec flashnet-trovi-env in /home/cc/.local/share/jupyter/kernels/flashnet-trovi-env\n",
      "8 =================================================================================\n",
      "Requirement already satisfied: jupyterthemes in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (0.20.0)\n",
      "Requirement already satisfied: jupyter-core in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (5.2.0)\n",
      "Requirement already satisfied: notebook>=5.6.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (6.5.2)\n",
      "Requirement already satisfied: ipython>=5.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (8.9.0)\n",
      "Requirement already satisfied: matplotlib>=1.4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (3.6.3)\n",
      "Requirement already satisfied: lesscpy>=0.11.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyterthemes) (0.15.1)\n",
      "Requirement already satisfied: matplotlib-inline in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.1.6)\n",
      "Requirement already satisfied: pickleshare in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.7.5)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (4.8.0)\n",
      "Requirement already satisfied: backcall in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.2.0)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.18.2)\n",
      "Requirement already satisfied: stack-data in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (0.6.2)\n",
      "Requirement already satisfied: traitlets>=5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (5.9.0)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.30 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (3.0.36)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (2.14.0)\n",
      "Requirement already satisfied: decorator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipython>=5.4.1->jupyterthemes) (5.1.1)\n",
      "Requirement already satisfied: ply in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from lesscpy>=0.11.2->jupyterthemes) (3.11)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (2.8.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (3.0.9)\n",
      "Requirement already satisfied: numpy>=1.19 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.24.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (0.11.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (23.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.0.7)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (9.4.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (1.4.4)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from matplotlib>=1.4.3->jupyterthemes) (4.38.0)\n",
      "Requirement already satisfied: pyzmq>=17 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (25.0.0)\n",
      "Requirement already satisfied: nbformat in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (5.7.3)\n",
      "Requirement already satisfied: nbclassic>=0.4.7 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.5.1)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (6.2)\n",
      "Requirement already satisfied: prometheus-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.16.0)\n",
      "Requirement already satisfied: nest-asyncio>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (1.5.6)\n",
      "Requirement already satisfied: jupyter-client>=5.3.4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (8.0.2)\n",
      "Requirement already satisfied: Send2Trash>=1.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (1.8.0)\n",
      "Requirement already satisfied: ipykernel in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (6.21.1)\n",
      "Requirement already satisfied: nbconvert>=5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (7.2.9)\n",
      "Requirement already satisfied: terminado>=0.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.17.1)\n",
      "Requirement already satisfied: argon2-cffi in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (21.3.0)\n",
      "Requirement already satisfied: ipython-genutils in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (0.2.0)\n",
      "Requirement already satisfied: jinja2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from notebook>=5.6.0->jupyterthemes) (3.1.2)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-core->jupyterthemes) (2.6.2)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jedi>=0.16->ipython>=5.4.1->jupyterthemes) (0.8.3)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-client>=5.3.4->notebook>=5.6.0->jupyterthemes) (6.0.0)\n",
      "Requirement already satisfied: jupyter-server>=1.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (2.2.1)\n",
      "Requirement already satisfied: notebook-shim>=0.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: defusedxml in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.7.1)\n",
      "Requirement already satisfied: markupsafe>=2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.1.2)\n",
      "Requirement already satisfied: bleach in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (6.0.0)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.7.2)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (1.2.1)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (4.11.2)\n",
      "Requirement already satisfied: mistune<3,>=2.0.3 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.0.4)\n",
      "Requirement already satisfied: jupyterlab-pygments in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: fastjsonschema in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbformat->notebook>=5.6.0->jupyterthemes) (2.16.2)\n",
      "Requirement already satisfied: jsonschema>=2.6 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from nbformat->notebook>=5.6.0->jupyterthemes) (4.17.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from pexpect>4.3->ipython>=5.4.1->jupyterthemes) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from prompt-toolkit<3.1.0,>=3.0.30->ipython>=5.4.1->jupyterthemes) (0.2.6)\n",
      "Requirement already satisfied: six>=1.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib>=1.4.3->jupyterthemes) (1.16.0)\n",
      "Requirement already satisfied: argon2-cffi-bindings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi->notebook>=5.6.0->jupyterthemes) (21.2.0)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (1.6.6)\n",
      "Requirement already satisfied: psutil in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (5.9.4)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from ipykernel->notebook>=5.6.0->jupyterthemes) (0.1.2)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (2.2.1)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (1.2.0)\n",
      "Requirement already satisfied: pure-eval in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from stack-data->ipython>=5.4.1->jupyterthemes) (0.2.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=5.3.4->notebook>=5.6.0->jupyterthemes) (3.12.0)\n",
      "Requirement already satisfied: pkgutil-resolve-name>=1.3.10 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.3.10)\n",
      "Requirement already satisfied: importlib-resources>=1.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (5.10.2)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (22.2.0)\n",
      "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.19.3)\n",
      "Requirement already satisfied: anyio>=3.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (3.6.2)\n",
      "Requirement already satisfied: jupyter-server-terminals in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.4.4)\n",
      "Requirement already satisfied: jupyter-events>=0.4.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (0.5.0)\n",
      "Requirement already satisfied: websocket-client in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (1.5.1)\n",
      "Requirement already satisfied: cffi>=1.0.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from argon2-cffi-bindings->argon2-cffi->notebook>=5.6.0->jupyterthemes) (1.15.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from beautifulsoup4->nbconvert>=5->notebook>=5.6.0->jupyterthemes) (2.3.2.post1)\n",
      "Requirement already satisfied: webencodings in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from bleach->nbconvert>=5->notebook>=5.6.0->jupyterthemes) (0.5.1)\n",
      "Requirement already satisfied: sniffio>=1.1 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (1.3.0)\n",
      "Requirement already satisfied: idna>=2.8 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from anyio>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (3.4)\n",
      "Requirement already satisfied: pycparser in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook>=5.6.0->jupyterthemes) (2.21)\n",
      "Requirement already satisfied: pyyaml in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (6.0)\n",
      "Requirement already satisfied: python-json-logger in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jupyter-events>=0.4.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook>=5.6.0->jupyterthemes) (2.0.4)\n",
      "Requirement already satisfied: jsonpointer>1.13 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (2.3)\n",
      "Requirement already satisfied: isoduration in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (20.11.0)\n",
      "Requirement already satisfied: fqdn in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.5.1)\n",
      "Requirement already satisfied: rfc3986-validator>0.1.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.1.1)\n",
      "Requirement already satisfied: webcolors>=1.11 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.12)\n",
      "Requirement already satisfied: rfc3339-validator in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (0.1.4)\n",
      "Requirement already satisfied: uri-template in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.2.0)\n",
      "Requirement already satisfied: arrow>=0.15.0 in /home/cc/anaconda3/envs/flashnet-trovi-env/lib/python3.8/site-packages (from isoduration->jsonschema>=2.6->nbformat->notebook>=5.6.0->jupyterthemes) (1.2.3)\n",
      "Available Themes: \n",
      "   chesterish\n",
      "   grade3\n",
      "   gruvboxd\n",
      "   gruvboxl\n",
      "   monokai\n",
      "   oceans16\n",
      "   onedork\n",
      "   solarizedd\n",
      "   solarizedl\n"
     ]
    }
   ],
   "source": [
    "# Creating conda env named \"flashnet-trovi-env\"\n",
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    cd flashnet-trovi\n",
    "    ./install_conda_deps.sh\n",
    "EOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cc/anaconda3/condabin/conda\n",
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... Solving environment: ...working... done\n",
      "done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 22.9.0\n",
      "  latest version: 23.1.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c defaults conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/cc/anaconda3/envs/flashnet-trovi-env\n",
      "\n",
      "  added / updated specs:\n",
      "    - _libgcc_mutex\n",
      "    - _openmp_mutex\n",
      "    - asttokens\n",
      "    - backcall\n",
      "    - ca-certificates\n",
      "    - certifi\n",
      "    - comm\n",
      "    - debugpy\n",
      "    - decorator\n",
      "    - entrypoints\n",
      "    - executing\n",
      "    - ipykernel\n",
      "    - ipython\n",
      "    - jedi\n",
      "    - jupyter_client\n",
      "    - jupyter_core\n",
      "    - ld_impl_linux-64\n",
      "    - libffi\n",
      "    - libgcc-ng\n",
      "    - libgomp\n",
      "    - libsodium\n",
      "    - libstdcxx-ng\n",
      "    - matplotlib-inline\n",
      "    - ncurses\n",
      "    - nest-asyncio\n",
      "    - openssl\n",
      "    - packaging\n",
      "    - parso\n",
      "    - pexpect\n",
      "    - pickleshare\n",
      "    - pip\n",
      "    - platformdirs\n",
      "    - prompt-toolkit\n",
      "    - psutil\n",
      "    - ptyprocess\n",
      "    - pure_eval\n",
      "    - pygments\n",
      "    - python-dateutil\n",
      "    - python=3.8\n",
      "    - pyzmq\n",
      "    - readline\n",
      "    - setuptools\n",
      "    - six\n",
      "    - sqlite\n",
      "    - stack_data\n",
      "    - tk\n",
      "    - tornado\n",
      "    - traitlets\n",
      "    - wcwidth\n",
      "    - wheel\n",
      "    - xz\n",
      "    - zeromq\n",
      "    - zlib\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  asttokens          pkgs/main/noarch::asttokens-2.0.5-pyhd3eb1b0_0 None\n",
      "  backcall           pkgs/main/noarch::backcall-0.2.0-pyhd3eb1b0_0 None\n",
      "  comm               pkgs/main/linux-64::comm-0.1.2-py38h06a4308_0 None\n",
      "  debugpy            pkgs/main/linux-64::debugpy-1.5.1-py38h295c915_0 None\n",
      "  decorator          pkgs/main/noarch::decorator-5.1.1-pyhd3eb1b0_0 None\n",
      "  entrypoints        pkgs/main/linux-64::entrypoints-0.4-py38h06a4308_0 None\n",
      "  executing          pkgs/main/noarch::executing-0.8.3-pyhd3eb1b0_0 None\n",
      "  ipykernel          pkgs/main/linux-64::ipykernel-6.19.2-py38hb070fc8_0 None\n",
      "  ipython            pkgs/main/linux-64::ipython-8.8.0-py38h06a4308_0 None\n",
      "  jedi               pkgs/main/linux-64::jedi-0.18.1-py38h06a4308_1 None\n",
      "  jupyter_client     pkgs/main/linux-64::jupyter_client-7.4.8-py38h06a4308_0 None\n",
      "  jupyter_core       pkgs/main/linux-64::jupyter_core-5.1.1-py38h06a4308_0 None\n",
      "  libsodium          pkgs/main/linux-64::libsodium-1.0.18-h7b6447c_0 None\n",
      "  matplotlib-inline  pkgs/main/linux-64::matplotlib-inline-0.1.6-py38h06a4308_0 None\n",
      "  nest-asyncio       pkgs/main/linux-64::nest-asyncio-1.5.6-py38h06a4308_0 None\n",
      "  packaging          pkgs/main/linux-64::packaging-22.0-py38h06a4308_0 None\n",
      "  parso              pkgs/main/noarch::parso-0.8.3-pyhd3eb1b0_0 None\n",
      "  pexpect            pkgs/main/noarch::pexpect-4.8.0-pyhd3eb1b0_3 None\n",
      "  pickleshare        pkgs/main/noarch::pickleshare-0.7.5-pyhd3eb1b0_1003 None\n",
      "  platformdirs       pkgs/main/linux-64::platformdirs-2.5.2-py38h06a4308_0 None\n",
      "  prompt-toolkit     pkgs/main/linux-64::prompt-toolkit-3.0.36-py38h06a4308_0 None\n",
      "  psutil             pkgs/main/linux-64::psutil-5.9.0-py38h5eee18b_0 None\n",
      "  ptyprocess         pkgs/main/noarch::ptyprocess-0.7.0-pyhd3eb1b0_2 None\n",
      "  pure_eval          pkgs/main/noarch::pure_eval-0.2.2-pyhd3eb1b0_0 None\n",
      "  pygments           pkgs/main/noarch::pygments-2.11.2-pyhd3eb1b0_0 None\n",
      "  python-dateutil    pkgs/main/noarch::python-dateutil-2.8.2-pyhd3eb1b0_0 None\n",
      "  pyzmq              pkgs/main/linux-64::pyzmq-23.2.0-py38h6a678d5_0 None\n",
      "  six                pkgs/main/noarch::six-1.16.0-pyhd3eb1b0_1 None\n",
      "  stack_data         pkgs/main/noarch::stack_data-0.2.0-pyhd3eb1b0_0 None\n",
      "  tornado            pkgs/main/linux-64::tornado-6.2-py38h5eee18b_0 None\n",
      "  traitlets          pkgs/main/linux-64::traitlets-5.7.1-py38h06a4308_0 None\n",
      "  wcwidth            pkgs/main/noarch::wcwidth-0.2.5-pyhd3eb1b0_0 None\n",
      "  zeromq             pkgs/main/linux-64::zeromq-4.3.4-h2531618_0 None\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "Retrieving notices: ...working... done\n"
     ]
    }
   ],
   "source": [
    "# Install ipykernel\n",
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    cd flashnet-trovi\n",
    "    source ~/.zshrc\n",
    "    which conda\n",
    "    conda install -n flashnet-trovi-env ipykernel --update-deps --force-reinstall -y\n",
    "EOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "FLASHNET=`eval \"$ssh_command\" /bin/bash << EOF\n",
    "    cd flashnet-trovi\n",
    "    pwd\n",
    "EOF`\n",
    "export FLASHNET=$FLASHNET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    export FLASHNET=$FLASHNET\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Run Tail Analyzer for Labeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trace_profiles = ['/home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.most_rand_iops.537.trace', '/home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.most_size_thpt.222.trace', '/home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_60_40.490.trace', '/home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_65_35.211.trace', '/home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_75_25.379.trace']\n",
      "\n",
      "Processing /home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.most_rand_iops.537.trace\n",
      "#IO labeled = 50000\n",
      "Fast IO = 36216\n",
      "Slow IO = 13784\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.labeled\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.lat_cdf.png\n",
      "\n",
      "Processing /home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.most_size_thpt.222.trace\n",
      "#IO labeled = 50000\n",
      "Fast IO = 38215\n",
      "Slow IO = 11785\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.labeled\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.lat_cdf.png\n",
      "\n",
      "Processing /home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_60_40.490.trace\n",
      "#IO labeled = 50000\n",
      "Fast IO = 34266\n",
      "Slow IO = 15734\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.labeled\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.lat_cdf.png\n",
      "\n",
      "Processing /home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_65_35.211.trace\n",
      "#IO labeled = 50000\n",
      "Fast IO = 34395\n",
      "Slow IO = 15605\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.labeled\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.lat_cdf.png\n",
      "\n",
      "Processing /home/cc/flashnet-trovi/data/trace_profile/nvme0n1/tencent.cut.per_50k.rw_75_25.379.trace\n",
      "#IO labeled = 50000\n",
      "Fast IO = 36260\n",
      "Slow IO = 13740\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.labeled\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.lat_cdf.png\n"
     ]
    }
   ],
   "source": [
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    source ~/.zshrc\n",
    "    cd $FLASHNET/model_collection/1_per_io_admission/tail_analyzer/\n",
    "    conda activate \"flashnet-trovi-env\" && ./tail_v2.py -files $FLASHNET/data/trace_profile/nvme0n1/tencent.cut.per_50k*.trace\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Run Feature Extractor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trace_profiles = ['../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.labeled', '../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.labeled', '../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.labeled', '../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.labeled', '../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.labeled']\n",
      "\n",
      "Processing ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.labeled\n",
      "Removed 3 first IOs because they don't have enough historical data\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.dataset\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Processing ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.labeled\n",
      "Removed 3 first IOs because they don't have enough historical data\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.dataset\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Processing ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.labeled\n",
      "Removed 3 first IOs because they don't have enough historical data\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.dataset\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Processing ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.labeled\n",
      "Removed 3 first IOs because they don't have enough historical data\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.dataset\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Processing ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.labeled\n",
      "Removed 3 first IOs because they don't have enough historical data\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.dataset\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.readonly.dataset\n"
     ]
    }
   ],
   "source": [
    "# run on multiple profiles\n",
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    source ~/.zshrc\n",
    "    cd $FLASHNET/model_collection/1_per_io_admission/feature_extractor/\n",
    "    conda activate flashnet-trovi-env && ./feat_v2.py -files ../dataset/nvme0n1/tencent.cut.per_50k*/profile_v2.labeled\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Train the NN model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-02-05 15:39:34.860515: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:39:34.860542: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:39:36.249793: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:39:36.249820: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:39:36.250051: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[8.2122582e-01 9.2154316e+03 1.4032305e+03 1.4001462e+03 1.4026096e+03\n",
      "  2.8238577e+01 2.8158209e+01 2.7460609e+01]]\n",
      "Epoch 1/20\n",
      "625/625 [==============================] - 4s 5ms/step - loss: 0.3097 - accuracy: 0.8923 - val_loss: 0.2559 - val_accuracy: 0.9044\n",
      "Epoch 2/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2620 - accuracy: 0.8968 - val_loss: 0.2440 - val_accuracy: 0.9070\n",
      "Epoch 3/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2535 - accuracy: 0.8992 - val_loss: 0.2377 - val_accuracy: 0.9070\n",
      "Epoch 4/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2486 - accuracy: 0.9007 - val_loss: 0.2350 - val_accuracy: 0.9076\n",
      "Epoch 5/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2443 - accuracy: 0.9020 - val_loss: 0.2321 - val_accuracy: 0.9094\n",
      "Epoch 6/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2419 - accuracy: 0.9021 - val_loss: 0.2334 - val_accuracy: 0.9090\n",
      "Epoch 7/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2392 - accuracy: 0.9046 - val_loss: 0.2303 - val_accuracy: 0.9120\n",
      "Epoch 8/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2382 - accuracy: 0.9057 - val_loss: 0.2285 - val_accuracy: 0.9110\n",
      "Epoch 9/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2364 - accuracy: 0.9060 - val_loss: 0.2282 - val_accuracy: 0.9128\n",
      "Epoch 10/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2355 - accuracy: 0.9065 - val_loss: 0.2263 - val_accuracy: 0.9106\n",
      "Epoch 11/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2338 - accuracy: 0.9071 - val_loss: 0.2276 - val_accuracy: 0.9100\n",
      "Epoch 12/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2330 - accuracy: 0.9076 - val_loss: 0.2274 - val_accuracy: 0.9110\n",
      "Epoch 13/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2325 - accuracy: 0.9078 - val_loss: 0.2244 - val_accuracy: 0.9120\n",
      "Epoch 14/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2317 - accuracy: 0.9085 - val_loss: 0.2248 - val_accuracy: 0.9122\n",
      "Epoch 15/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2309 - accuracy: 0.9092 - val_loss: 0.2238 - val_accuracy: 0.9144\n",
      "Epoch 16/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2305 - accuracy: 0.9110 - val_loss: 0.2237 - val_accuracy: 0.9138\n",
      "Epoch 17/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2298 - accuracy: 0.9099 - val_loss: 0.2254 - val_accuracy: 0.9138\n",
      "Epoch 18/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2293 - accuracy: 0.9100 - val_loss: 0.2256 - val_accuracy: 0.9130\n",
      "Epoch 19/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2288 - accuracy: 0.9108 - val_loss: 0.2224 - val_accuracy: 0.9130\n",
      "Epoch 20/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2279 - accuracy: 0.9113 - val_loss: 0.2261 - val_accuracy: 0.9132\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2/model_binary_nn/eval.png\n",
      "2023-02-05 15:40:42.998977: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:40:42.999004: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:40:44.383123: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:40:44.383151: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:40:44.383381: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[1.00000000e+00 8.65139453e+03 1.55035022e+03 1.54121460e+03\n",
      "  1.53899670e+03 1.40773945e+01 1.56729841e+01 1.61304436e+01]]\n",
      "Epoch 1/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.3562 - accuracy: 0.8716 - val_loss: 0.3196 - val_accuracy: 0.8777\n",
      "Epoch 2/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.3085 - accuracy: 0.8788 - val_loss: 0.3118 - val_accuracy: 0.8779\n",
      "Epoch 3/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.3010 - accuracy: 0.8800 - val_loss: 0.3040 - val_accuracy: 0.8811\n",
      "Epoch 4/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2943 - accuracy: 0.8806 - val_loss: 0.3002 - val_accuracy: 0.8830\n",
      "Epoch 5/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2893 - accuracy: 0.8859 - val_loss: 0.2972 - val_accuracy: 0.8852\n",
      "Epoch 6/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2854 - accuracy: 0.8858 - val_loss: 0.2925 - val_accuracy: 0.8869\n",
      "Epoch 7/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2823 - accuracy: 0.8887 - val_loss: 0.2904 - val_accuracy: 0.8855\n",
      "Epoch 8/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2795 - accuracy: 0.8900 - val_loss: 0.2887 - val_accuracy: 0.8847\n",
      "Epoch 9/20\n",
      "513/513 [==============================] - 2s 5ms/step - loss: 0.2773 - accuracy: 0.8888 - val_loss: 0.2904 - val_accuracy: 0.8813\n",
      "Epoch 10/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2760 - accuracy: 0.8917 - val_loss: 0.2887 - val_accuracy: 0.8830\n",
      "Epoch 11/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2740 - accuracy: 0.8912 - val_loss: 0.2874 - val_accuracy: 0.8882\n",
      "Epoch 12/20\n",
      "513/513 [==============================] - 2s 5ms/step - loss: 0.2734 - accuracy: 0.8911 - val_loss: 0.2857 - val_accuracy: 0.8857\n",
      "Epoch 13/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2719 - accuracy: 0.8925 - val_loss: 0.2848 - val_accuracy: 0.8823\n",
      "Epoch 14/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2707 - accuracy: 0.8937 - val_loss: 0.2876 - val_accuracy: 0.8799\n",
      "Epoch 15/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2702 - accuracy: 0.8923 - val_loss: 0.2841 - val_accuracy: 0.8874\n",
      "Epoch 16/20\n",
      "513/513 [==============================] - 2s 5ms/step - loss: 0.2690 - accuracy: 0.8950 - val_loss: 0.2831 - val_accuracy: 0.8874\n",
      "Epoch 17/20\n",
      "513/513 [==============================] - 2s 5ms/step - loss: 0.2684 - accuracy: 0.8942 - val_loss: 0.2829 - val_accuracy: 0.8850\n",
      "Epoch 18/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2678 - accuracy: 0.8942 - val_loss: 0.2843 - val_accuracy: 0.8855\n",
      "Epoch 19/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2676 - accuracy: 0.8949 - val_loss: 0.2820 - val_accuracy: 0.8845\n",
      "Epoch 20/20\n",
      "513/513 [==============================] - 3s 5ms/step - loss: 0.2660 - accuracy: 0.8953 - val_loss: 0.2815 - val_accuracy: 0.8855\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.readonly/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.readonly/model_binary_nn/eval.png\n",
      "2023-02-05 15:41:39.625083: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:41:39.625111: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:41:41.010143: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:41:41.010171: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:41:41.010402: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[8.31826389e-01 2.56538047e+04 1.91917346e+03 1.91786145e+03\n",
      "  1.92158630e+03 1.59858665e+01 1.56592913e+01 1.56299715e+01]]\n",
      "Epoch 1/20\n",
      "625/625 [==============================] - 4s 5ms/step - loss: 0.2624 - accuracy: 0.9082 - val_loss: 0.2133 - val_accuracy: 0.9162\n",
      "Epoch 2/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2090 - accuracy: 0.9198 - val_loss: 0.2001 - val_accuracy: 0.9220\n",
      "Epoch 3/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2011 - accuracy: 0.9232 - val_loss: 0.1966 - val_accuracy: 0.9222\n",
      "Epoch 4/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1979 - accuracy: 0.9235 - val_loss: 0.1950 - val_accuracy: 0.9222\n",
      "Epoch 5/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1956 - accuracy: 0.9251 - val_loss: 0.1947 - val_accuracy: 0.9226\n",
      "Epoch 6/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1940 - accuracy: 0.9238 - val_loss: 0.1926 - val_accuracy: 0.9206\n",
      "Epoch 7/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1924 - accuracy: 0.9255 - val_loss: 0.1892 - val_accuracy: 0.9234\n",
      "Epoch 8/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1909 - accuracy: 0.9262 - val_loss: 0.1880 - val_accuracy: 0.9232\n",
      "Epoch 9/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1898 - accuracy: 0.9262 - val_loss: 0.1876 - val_accuracy: 0.9216\n",
      "Epoch 10/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1887 - accuracy: 0.9261 - val_loss: 0.1868 - val_accuracy: 0.9236\n",
      "Epoch 11/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1882 - accuracy: 0.9266 - val_loss: 0.1849 - val_accuracy: 0.9244\n",
      "Epoch 12/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1871 - accuracy: 0.9273 - val_loss: 0.1864 - val_accuracy: 0.9242\n",
      "Epoch 13/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1864 - accuracy: 0.9270 - val_loss: 0.1865 - val_accuracy: 0.9226\n",
      "Epoch 14/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1858 - accuracy: 0.9276 - val_loss: 0.1837 - val_accuracy: 0.9234\n",
      "Epoch 15/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1854 - accuracy: 0.9279 - val_loss: 0.1832 - val_accuracy: 0.9250\n",
      "Epoch 16/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1846 - accuracy: 0.9273 - val_loss: 0.1848 - val_accuracy: 0.9244\n",
      "Epoch 17/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1840 - accuracy: 0.9285 - val_loss: 0.1835 - val_accuracy: 0.9238\n",
      "Epoch 18/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1837 - accuracy: 0.9281 - val_loss: 0.1828 - val_accuracy: 0.9270\n",
      "Epoch 19/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1836 - accuracy: 0.9273 - val_loss: 0.1826 - val_accuracy: 0.9262\n",
      "Epoch 20/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.1830 - accuracy: 0.9284 - val_loss: 0.1824 - val_accuracy: 0.9248\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2/model_binary_nn/eval.png\n",
      "2023-02-05 15:42:47.807859: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:42:47.807888: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:42:49.189784: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:42:49.189812: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:42:49.190038: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[1.0000000e+00 2.9571211e+04 2.1310269e+03 2.1247959e+03 2.1238579e+03\n",
      "  1.5582695e+01 1.5502524e+01 1.5549119e+01]]\n",
      "Epoch 1/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2910 - accuracy: 0.8917 - val_loss: 0.2519 - val_accuracy: 0.8977\n",
      "Epoch 2/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2425 - accuracy: 0.9078 - val_loss: 0.2467 - val_accuracy: 0.9011\n",
      "Epoch 3/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2363 - accuracy: 0.9079 - val_loss: 0.2402 - val_accuracy: 0.9059\n",
      "Epoch 4/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2336 - accuracy: 0.9086 - val_loss: 0.2424 - val_accuracy: 0.9068\n",
      "Epoch 5/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2310 - accuracy: 0.9104 - val_loss: 0.2390 - val_accuracy: 0.9037\n",
      "Epoch 6/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2292 - accuracy: 0.9102 - val_loss: 0.2381 - val_accuracy: 0.9037\n",
      "Epoch 7/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2279 - accuracy: 0.9097 - val_loss: 0.2354 - val_accuracy: 0.9068\n",
      "Epoch 8/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2261 - accuracy: 0.9109 - val_loss: 0.2338 - val_accuracy: 0.9085\n",
      "Epoch 9/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2248 - accuracy: 0.9096 - val_loss: 0.2332 - val_accuracy: 0.9075\n",
      "Epoch 10/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2234 - accuracy: 0.9120 - val_loss: 0.2327 - val_accuracy: 0.9080\n",
      "Epoch 11/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2224 - accuracy: 0.9113 - val_loss: 0.2313 - val_accuracy: 0.9083\n",
      "Epoch 12/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2215 - accuracy: 0.9112 - val_loss: 0.2305 - val_accuracy: 0.9083\n",
      "Epoch 13/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2201 - accuracy: 0.9118 - val_loss: 0.2323 - val_accuracy: 0.9071\n",
      "Epoch 14/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2186 - accuracy: 0.9116 - val_loss: 0.2296 - val_accuracy: 0.9068\n",
      "Epoch 15/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2190 - accuracy: 0.9120 - val_loss: 0.2319 - val_accuracy: 0.9075\n",
      "Epoch 16/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2177 - accuracy: 0.9133 - val_loss: 0.2289 - val_accuracy: 0.9090\n",
      "Epoch 17/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2176 - accuracy: 0.9132 - val_loss: 0.2306 - val_accuracy: 0.9090\n",
      "Epoch 18/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2164 - accuracy: 0.9145 - val_loss: 0.2315 - val_accuracy: 0.9097\n",
      "Epoch 19/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2167 - accuracy: 0.9130 - val_loss: 0.2282 - val_accuracy: 0.9087\n",
      "Epoch 20/20\n",
      "521/521 [==============================] - 3s 5ms/step - loss: 0.2158 - accuracy: 0.9141 - val_loss: 0.2290 - val_accuracy: 0.9066\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.readonly/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.readonly/model_binary_nn/eval.png\n",
      "2023-02-05 15:43:45.551814: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:43:45.551843: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:43:46.939786: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:43:46.939815: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:43:46.940031: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[6.0012865e-01 1.2135454e+04 1.1996090e+03 1.2010928e+03 1.2041250e+03\n",
      "  4.1297359e+01 4.1690647e+01 4.1668209e+01]]\n",
      "Epoch 1/20\n",
      "625/625 [==============================] - 4s 5ms/step - loss: 0.2946 - accuracy: 0.8844 - val_loss: 0.2536 - val_accuracy: 0.8916\n",
      "Epoch 2/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2527 - accuracy: 0.8911 - val_loss: 0.2541 - val_accuracy: 0.8884\n",
      "Epoch 3/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2489 - accuracy: 0.8917 - val_loss: 0.2477 - val_accuracy: 0.8910\n",
      "Epoch 4/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2467 - accuracy: 0.8919 - val_loss: 0.2454 - val_accuracy: 0.8934\n",
      "Epoch 5/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2445 - accuracy: 0.8933 - val_loss: 0.2442 - val_accuracy: 0.8938\n",
      "Epoch 6/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2432 - accuracy: 0.8934 - val_loss: 0.2441 - val_accuracy: 0.8946\n",
      "Epoch 7/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2419 - accuracy: 0.8946 - val_loss: 0.2436 - val_accuracy: 0.8964\n",
      "Epoch 8/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2406 - accuracy: 0.8953 - val_loss: 0.2438 - val_accuracy: 0.8974\n",
      "Epoch 9/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2398 - accuracy: 0.8953 - val_loss: 0.2420 - val_accuracy: 0.8958\n",
      "Epoch 10/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2386 - accuracy: 0.8958 - val_loss: 0.2405 - val_accuracy: 0.8970\n",
      "Epoch 11/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2381 - accuracy: 0.8962 - val_loss: 0.2385 - val_accuracy: 0.8952\n",
      "Epoch 12/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2369 - accuracy: 0.8969 - val_loss: 0.2386 - val_accuracy: 0.8960\n",
      "Epoch 13/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2367 - accuracy: 0.8981 - val_loss: 0.2395 - val_accuracy: 0.8966\n",
      "Epoch 14/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2357 - accuracy: 0.8959 - val_loss: 0.2380 - val_accuracy: 0.8936\n",
      "Epoch 15/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2351 - accuracy: 0.8969 - val_loss: 0.2387 - val_accuracy: 0.8940\n",
      "Epoch 16/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2345 - accuracy: 0.8977 - val_loss: 0.2378 - val_accuracy: 0.8946\n",
      "Epoch 17/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2339 - accuracy: 0.8977 - val_loss: 0.2370 - val_accuracy: 0.8934\n",
      "Epoch 18/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2336 - accuracy: 0.8982 - val_loss: 0.2357 - val_accuracy: 0.8954\n",
      "Epoch 19/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2337 - accuracy: 0.8977 - val_loss: 0.2361 - val_accuracy: 0.8974\n",
      "Epoch 20/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2328 - accuracy: 0.8988 - val_loss: 0.2363 - val_accuracy: 0.8964\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2/model_binary_nn/eval.png\n",
      "2023-02-05 15:44:54.379285: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:44:54.379312: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:44:55.748598: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:44:55.748626: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:44:55.748849: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[1.0000000e+00 1.1951129e+04 1.6033530e+03 1.5920044e+03 1.5681174e+03\n",
      "  1.1676747e+01 1.5243935e+01 1.5015527e+01]]\n",
      "Epoch 1/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4612 - accuracy: 0.8080 - val_loss: 0.4333 - val_accuracy: 0.8167\n",
      "Epoch 2/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4241 - accuracy: 0.8193 - val_loss: 0.4233 - val_accuracy: 0.8147\n",
      "Epoch 3/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4169 - accuracy: 0.8204 - val_loss: 0.4189 - val_accuracy: 0.8174\n",
      "Epoch 4/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4118 - accuracy: 0.8217 - val_loss: 0.4157 - val_accuracy: 0.8194\n",
      "Epoch 5/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4082 - accuracy: 0.8240 - val_loss: 0.4077 - val_accuracy: 0.8167\n",
      "Epoch 6/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4051 - accuracy: 0.8232 - val_loss: 0.4053 - val_accuracy: 0.8211\n",
      "Epoch 7/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4023 - accuracy: 0.8252 - val_loss: 0.4050 - val_accuracy: 0.8217\n",
      "Epoch 8/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.4007 - accuracy: 0.8266 - val_loss: 0.4029 - val_accuracy: 0.8231\n",
      "Epoch 9/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3984 - accuracy: 0.8287 - val_loss: 0.4044 - val_accuracy: 0.8214\n",
      "Epoch 10/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3968 - accuracy: 0.8255 - val_loss: 0.4016 - val_accuracy: 0.8247\n",
      "Epoch 11/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3952 - accuracy: 0.8283 - val_loss: 0.4001 - val_accuracy: 0.8207\n",
      "Epoch 12/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3939 - accuracy: 0.8291 - val_loss: 0.4011 - val_accuracy: 0.8197\n",
      "Epoch 13/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3915 - accuracy: 0.8289 - val_loss: 0.3974 - val_accuracy: 0.8211\n",
      "Epoch 14/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3922 - accuracy: 0.8284 - val_loss: 0.3985 - val_accuracy: 0.8264\n",
      "Epoch 15/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3910 - accuracy: 0.8288 - val_loss: 0.3983 - val_accuracy: 0.8257\n",
      "Epoch 16/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3894 - accuracy: 0.8301 - val_loss: 0.3969 - val_accuracy: 0.8264\n",
      "Epoch 17/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3885 - accuracy: 0.8301 - val_loss: 0.3964 - val_accuracy: 0.8241\n",
      "Epoch 18/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3880 - accuracy: 0.8310 - val_loss: 0.3961 - val_accuracy: 0.8231\n",
      "Epoch 19/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3869 - accuracy: 0.8305 - val_loss: 0.3956 - val_accuracy: 0.8244\n",
      "Epoch 20/20\n",
      "376/376 [==============================] - 2s 5ms/step - loss: 0.3861 - accuracy: 0.8316 - val_loss: 0.3937 - val_accuracy: 0.8247\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.readonly/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.readonly/model_binary_nn/eval.png\n",
      "2023-02-05 15:45:37.481099: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:45:37.481127: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:45:38.849652: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:45:38.849679: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:45:38.849908: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[6.5693301e-01 1.1375179e+04 1.2390829e+03 1.2346439e+03 1.2344220e+03\n",
      "  3.9443455e+01 3.9640114e+01 3.9791630e+01]]\n",
      "Epoch 1/20\n",
      "625/625 [==============================] - 4s 5ms/step - loss: 0.3037 - accuracy: 0.8821 - val_loss: 0.2690 - val_accuracy: 0.8860\n",
      "Epoch 2/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2625 - accuracy: 0.8873 - val_loss: 0.2634 - val_accuracy: 0.8858\n",
      "Epoch 3/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2577 - accuracy: 0.8889 - val_loss: 0.2584 - val_accuracy: 0.8916\n",
      "Epoch 4/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2538 - accuracy: 0.8900 - val_loss: 0.2564 - val_accuracy: 0.8900\n",
      "Epoch 5/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2503 - accuracy: 0.8911 - val_loss: 0.2547 - val_accuracy: 0.8906\n",
      "Epoch 6/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2489 - accuracy: 0.8922 - val_loss: 0.2530 - val_accuracy: 0.8906\n",
      "Epoch 7/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2470 - accuracy: 0.8923 - val_loss: 0.2503 - val_accuracy: 0.8932\n",
      "Epoch 8/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2450 - accuracy: 0.8942 - val_loss: 0.2506 - val_accuracy: 0.8952\n",
      "Epoch 9/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2437 - accuracy: 0.8934 - val_loss: 0.2487 - val_accuracy: 0.8946\n",
      "Epoch 10/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2424 - accuracy: 0.8952 - val_loss: 0.2487 - val_accuracy: 0.8914\n",
      "Epoch 11/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2414 - accuracy: 0.8960 - val_loss: 0.2480 - val_accuracy: 0.8966\n",
      "Epoch 12/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2408 - accuracy: 0.8953 - val_loss: 0.2473 - val_accuracy: 0.8942\n",
      "Epoch 13/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2403 - accuracy: 0.8957 - val_loss: 0.2489 - val_accuracy: 0.8916\n",
      "Epoch 14/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2397 - accuracy: 0.8963 - val_loss: 0.2475 - val_accuracy: 0.8922\n",
      "Epoch 15/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2391 - accuracy: 0.8953 - val_loss: 0.2479 - val_accuracy: 0.8964\n",
      "Epoch 16/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2389 - accuracy: 0.8969 - val_loss: 0.2477 - val_accuracy: 0.8948\n",
      "Epoch 17/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2380 - accuracy: 0.8958 - val_loss: 0.2469 - val_accuracy: 0.8900\n",
      "Epoch 18/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2379 - accuracy: 0.8967 - val_loss: 0.2504 - val_accuracy: 0.8902\n",
      "Epoch 19/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2368 - accuracy: 0.8971 - val_loss: 0.2467 - val_accuracy: 0.8938\n",
      "Epoch 20/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2371 - accuracy: 0.8983 - val_loss: 0.2477 - val_accuracy: 0.8928\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2/model_binary_nn/eval.png\n",
      "2023-02-05 15:46:44.901613: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:46:44.901638: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:46:46.269596: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:46:46.269623: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:46:46.269854: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[1.0000000e+00 1.0808831e+04 1.5907959e+03 1.5850159e+03 1.5640159e+03\n",
      "  1.1379814e+01 1.5438192e+01 1.5376787e+01]]\n",
      "Epoch 1/20\n",
      "412/412 [==============================] - 3s 5ms/step - loss: 0.4399 - accuracy: 0.8274 - val_loss: 0.4167 - val_accuracy: 0.8237\n",
      "Epoch 2/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.4022 - accuracy: 0.8334 - val_loss: 0.4086 - val_accuracy: 0.8237\n",
      "Epoch 3/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3935 - accuracy: 0.8357 - val_loss: 0.3975 - val_accuracy: 0.8204\n",
      "Epoch 4/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3878 - accuracy: 0.8370 - val_loss: 0.3933 - val_accuracy: 0.8219\n",
      "Epoch 5/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3822 - accuracy: 0.8395 - val_loss: 0.3913 - val_accuracy: 0.8216\n",
      "Epoch 6/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3786 - accuracy: 0.8384 - val_loss: 0.3904 - val_accuracy: 0.8222\n",
      "Epoch 7/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3763 - accuracy: 0.8402 - val_loss: 0.3872 - val_accuracy: 0.8237\n",
      "Epoch 8/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3741 - accuracy: 0.8409 - val_loss: 0.3839 - val_accuracy: 0.8243\n",
      "Epoch 9/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3727 - accuracy: 0.8418 - val_loss: 0.3817 - val_accuracy: 0.8213\n",
      "Epoch 10/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3702 - accuracy: 0.8442 - val_loss: 0.3853 - val_accuracy: 0.8265\n",
      "Epoch 11/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3685 - accuracy: 0.8442 - val_loss: 0.3839 - val_accuracy: 0.8240\n",
      "Epoch 12/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3685 - accuracy: 0.8440 - val_loss: 0.3801 - val_accuracy: 0.8289\n",
      "Epoch 13/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3664 - accuracy: 0.8454 - val_loss: 0.3789 - val_accuracy: 0.8243\n",
      "Epoch 14/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3658 - accuracy: 0.8460 - val_loss: 0.3808 - val_accuracy: 0.8246\n",
      "Epoch 15/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3651 - accuracy: 0.8450 - val_loss: 0.3816 - val_accuracy: 0.8246\n",
      "Epoch 16/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3636 - accuracy: 0.8478 - val_loss: 0.3807 - val_accuracy: 0.8234\n",
      "Epoch 17/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3628 - accuracy: 0.8457 - val_loss: 0.3769 - val_accuracy: 0.8262\n",
      "Epoch 18/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3628 - accuracy: 0.8460 - val_loss: 0.3816 - val_accuracy: 0.8271\n",
      "Epoch 19/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3621 - accuracy: 0.8468 - val_loss: 0.3815 - val_accuracy: 0.8246\n",
      "Epoch 20/20\n",
      "412/412 [==============================] - 2s 5ms/step - loss: 0.3614 - accuracy: 0.8460 - val_loss: 0.3778 - val_accuracy: 0.8277\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.readonly/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.readonly/model_binary_nn/eval.png\n",
      "2023-02-05 15:47:31.859505: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:47:31.859529: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:47:33.268370: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:47:33.268397: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:47:33.268607: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[7.5134057e-01 1.2798403e+04 1.4172970e+03 1.4208021e+03 1.4154316e+03\n",
      "  3.3448116e+01 3.3381634e+01 3.2819305e+01]]\n",
      "Epoch 1/20\n",
      "625/625 [==============================] - 4s 5ms/step - loss: 0.3010 - accuracy: 0.8897 - val_loss: 0.2573 - val_accuracy: 0.9000\n",
      "Epoch 2/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2532 - accuracy: 0.8978 - val_loss: 0.2487 - val_accuracy: 0.9006\n",
      "Epoch 3/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2463 - accuracy: 0.8992 - val_loss: 0.2431 - val_accuracy: 0.9032\n",
      "Epoch 4/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2407 - accuracy: 0.9019 - val_loss: 0.2367 - val_accuracy: 0.9040\n",
      "Epoch 5/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2371 - accuracy: 0.9038 - val_loss: 0.2328 - val_accuracy: 0.9062\n",
      "Epoch 6/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2343 - accuracy: 0.9054 - val_loss: 0.2315 - val_accuracy: 0.9072\n",
      "Epoch 7/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2318 - accuracy: 0.9049 - val_loss: 0.2284 - val_accuracy: 0.9084\n",
      "Epoch 8/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2302 - accuracy: 0.9054 - val_loss: 0.2302 - val_accuracy: 0.9092\n",
      "Epoch 9/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2288 - accuracy: 0.9066 - val_loss: 0.2268 - val_accuracy: 0.9084\n",
      "Epoch 10/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2283 - accuracy: 0.9044 - val_loss: 0.2258 - val_accuracy: 0.9076\n",
      "Epoch 11/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2269 - accuracy: 0.9063 - val_loss: 0.2282 - val_accuracy: 0.9078\n",
      "Epoch 12/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2259 - accuracy: 0.9084 - val_loss: 0.2278 - val_accuracy: 0.9078\n",
      "Epoch 13/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2251 - accuracy: 0.9078 - val_loss: 0.2245 - val_accuracy: 0.9092\n",
      "Epoch 14/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2248 - accuracy: 0.9077 - val_loss: 0.2239 - val_accuracy: 0.9096\n",
      "Epoch 15/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2245 - accuracy: 0.9073 - val_loss: 0.2257 - val_accuracy: 0.9104\n",
      "Epoch 16/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2236 - accuracy: 0.9085 - val_loss: 0.2267 - val_accuracy: 0.9074\n",
      "Epoch 17/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2234 - accuracy: 0.9084 - val_loss: 0.2231 - val_accuracy: 0.9114\n",
      "Epoch 18/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2227 - accuracy: 0.9089 - val_loss: 0.2232 - val_accuracy: 0.9084\n",
      "Epoch 19/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2219 - accuracy: 0.9094 - val_loss: 0.2235 - val_accuracy: 0.9104\n",
      "Epoch 20/20\n",
      "625/625 [==============================] - 3s 5ms/step - loss: 0.2219 - accuracy: 0.9107 - val_loss: 0.2233 - val_accuracy: 0.9112\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2/model_binary_nn/eval.png\n",
      "2023-02-05 15:48:39.885314: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-02-05 15:48:39.885341: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-02-05 15:48:41.268488: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2023-02-05 15:48:41.268519: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (rani-flashnet4): /proc/driver/nvidia/version does not exist\n",
      "2023-02-05 15:48:41.268781: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "[[1.0000000e+00 1.3368988e+04 1.6702443e+03 1.6627412e+03 1.6439240e+03\n",
      "  1.2688361e+01 1.5417907e+01 1.5368132e+01]]\n",
      "Epoch 1/20\n",
      "474/474 [==============================] - 3s 5ms/step - loss: 0.3754 - accuracy: 0.8629 - val_loss: 0.3466 - val_accuracy: 0.8588\n",
      "Epoch 2/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3303 - accuracy: 0.8711 - val_loss: 0.3395 - val_accuracy: 0.8646\n",
      "Epoch 3/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3225 - accuracy: 0.8735 - val_loss: 0.3358 - val_accuracy: 0.8665\n",
      "Epoch 4/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3169 - accuracy: 0.8744 - val_loss: 0.3274 - val_accuracy: 0.8683\n",
      "Epoch 5/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3106 - accuracy: 0.8778 - val_loss: 0.3251 - val_accuracy: 0.8675\n",
      "Epoch 6/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3071 - accuracy: 0.8779 - val_loss: 0.3205 - val_accuracy: 0.8688\n",
      "Epoch 7/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3027 - accuracy: 0.8785 - val_loss: 0.3217 - val_accuracy: 0.8675\n",
      "Epoch 8/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.3000 - accuracy: 0.8803 - val_loss: 0.3188 - val_accuracy: 0.8675\n",
      "Epoch 9/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2979 - accuracy: 0.8806 - val_loss: 0.3144 - val_accuracy: 0.8680\n",
      "Epoch 10/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2951 - accuracy: 0.8802 - val_loss: 0.3175 - val_accuracy: 0.8683\n",
      "Epoch 11/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2941 - accuracy: 0.8822 - val_loss: 0.3112 - val_accuracy: 0.8688\n",
      "Epoch 12/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2925 - accuracy: 0.8820 - val_loss: 0.3108 - val_accuracy: 0.8728\n",
      "Epoch 13/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2908 - accuracy: 0.8834 - val_loss: 0.3122 - val_accuracy: 0.8715\n",
      "Epoch 14/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2901 - accuracy: 0.8836 - val_loss: 0.3113 - val_accuracy: 0.8725\n",
      "Epoch 15/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2887 - accuracy: 0.8846 - val_loss: 0.3098 - val_accuracy: 0.8728\n",
      "Epoch 16/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2882 - accuracy: 0.8846 - val_loss: 0.3112 - val_accuracy: 0.8696\n",
      "Epoch 17/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2870 - accuracy: 0.8839 - val_loss: 0.3109 - val_accuracy: 0.8709\n",
      "Epoch 18/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2865 - accuracy: 0.8842 - val_loss: 0.3095 - val_accuracy: 0.8733\n",
      "Epoch 19/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2858 - accuracy: 0.8827 - val_loss: 0.3076 - val_accuracy: 0.8712\n",
      "Epoch 20/20\n",
      "474/474 [==============================] - 2s 5ms/step - loss: 0.2853 - accuracy: 0.8844 - val_loss: 0.3065 - val_accuracy: 0.8744\n",
      "===== output file : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.readonly/model_binary_nn/eval.stats\n",
      "===== output figure : ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.readonly/model_binary_nn/eval.png\n",
      "trace_profiles = ['../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.readonly.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.readonly.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.readonly.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.readonly.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.dataset', '../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.readonly.dataset']\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.most_rand_iops.537/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.most_size_thpt.222/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_60_40.490/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_65_35.211/profile_v2.feat_v2.readonly.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.dataset\n",
      "\n",
      "Training on ../dataset/nvme0n1/tencent.cut.per_50k.rw_75_25.379/profile_v2.feat_v2.readonly.dataset\n"
     ]
    }
   ],
   "source": [
    "# train on multiple datasets\n",
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    source ~/.zshrc\n",
    "    cd $FLASHNET/model_collection/1_per_io_admission/train/\n",
    "    conda activate flashnet-trovi-env && ./train_and_eval.py -model model_binary_nn -datasets ../dataset/nvme0n1/tencent*cut*per*/profile*feat*.dataset -train_eval_split 50_50\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Analyze the model performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 10 stats files\n",
      "===== output file : ../dataset/models_performance.csv\n"
     ]
    }
   ],
   "source": [
    "# First, we will gather all the stats\n",
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "    cd $FLASHNET/model_collection/1_per_io_admission/script/\n",
    "    source ~/.zshrc\n",
    "    conda activate flashnet-trovi-env && ./gather_eval_stats.py -files ../dataset/nvme*/*cut*/profile*/*/eval.stats\n",
    "EOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "eval \"$ssh_command\" /bin/bash << EOF\n",
    "source ~/.zshrc\n",
    "cd flashnet-trovi\n",
    "python3 -c '\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "df = pd.read_csv(\"model_collection/1_per_io_admission/dataset/models_performance.csv\")\n",
    "df.head()\n",
    "\n",
    "def plot_bars(df, column):\n",
    "    # Plot the ROC-AUC value from the non read_only dataset\n",
    "    df = df[df[\"read_only\"] == False]\n",
    "    fig, ax = plt.subplots(figsize=(5, 4))\n",
    "    # creating the bar plot\n",
    "    plt.bar(df[\"trace_name\"].tolist(), df[column].tolist(), color =[np.random.rand(3,) for x in range(len(df))], width = 0.4)\n",
    "    plt.setp( ax.xaxis.get_majorticklabels(), rotation=30, ha=\"right\" )\n",
    "    # ax.yaxis.set_ticks(np.arange(0, 1.1, 0.2))\n",
    "    plt.xlabel(\"Dataset name\")\n",
    "    plt.ylabel(column.upper())\n",
    "    plt.title(column.upper() + \" on Various Dataset\")\n",
    "    # plt.ylim(0,1)\n",
    "    plt.show()\n",
    "    plt.savefig(column+\".png\")\n",
    "\n",
    "plot_bars(df, \"roc_auc\")\n",
    "plot_bars(df, \"fnr\")\n",
    "plot_bars(df, \"fpr\")'\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View output in $FLASHNET"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "vscode": {
   "interpreter": {
    "hash": "3629df08e664f4d12bf036dd1ccc650b666adc62c0439f614bae21c1503e013e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
